{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2874521-df5c-4dd4-a334-1d26287882e6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 309 entries, 0 to 308\n",
      "Data columns (total 26 columns):\n",
      " #   Column                    Non-Null Count  Dtype \n",
      "---  ------                    --------------  ----- \n",
      " 0   ID_CNIO                   309 non-null    object\n",
      " 1   HISTOLOGY                 306 non-null    object\n",
      " 2   AGE                       309 non-null    object\n",
      " 3   GRADE                     243 non-null    object\n",
      " 4   FIGO                      275 non-null    object\n",
      " 5   FIGOL                     276 non-null    object\n",
      " 6   FIGOa                     275 non-null    object\n",
      " 7   NEOADJUVANT_TREATMENT     233 non-null    object\n",
      " 8   ADJUVANT_TREATMENT        279 non-null    object\n",
      " 9   TYPE_ADJUVANT             197 non-null    object\n",
      " 10  RESIDUAL                  224 non-null    object\n",
      " 11  RESIDUALa                 224 non-null    object\n",
      " 12  OS_CNIO                   284 non-null    object\n",
      " 13  VITAL STATUS              289 non-null    object\n",
      " 14  MMR STATUS                179 non-null    object\n",
      " 15  BRCA STATUS               169 non-null    object\n",
      " 16  FAMILIAL                  266 non-null    object\n",
      " 17  DATE_OS                   308 non-null    object\n",
      " 18  ID_ORIGINAL               299 non-null    object\n",
      " 19  ID_ORIGINAL_NHC_BBANK     283 non-null    object\n",
      " 20  OTHER_ADJ_TREAT           14 non-null     object\n",
      " 21  PARTIAL DATE DEATH_LASTv  61 non-null     object\n",
      " 22  PARTIAL DATE DIAGNOSIS    61 non-null     object\n",
      " 23  HISTOLOGY_DETAIL          129 non-null    object\n",
      " 24  DATE_AGE                  129 non-null    object\n",
      " 25  Clinical_info             309 non-null    object\n",
      "dtypes: object(26)\n",
      "memory usage: 62.9+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "### COMBINE ALL TYPE OF DATA IN A FINAL DATABASE AND OTHER DATABASES WITH SPECIFIC INFORMATION ############\n",
    "\n",
    "## Import neccesary libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import os\n",
    "### Read and combine final clinical databases\n",
    "# List of files to combine\n",
    "files = ['LP_final.xlsx', 'MDA_final.xlsx', 'OVE_final.xlsx', 'RVB_final.xlsx']\n",
    "# Read and combine eliminating rows of descriptions and coding (index 0 and 1). There are not the same in the different files so it is better to \n",
    "# create them again when we have all columns in the final order in the definitive file\n",
    "dataframes = []\n",
    "for file in files:\n",
    "    df = pd.read_excel(file)\n",
    "    df = df.drop([0,1]) # Eliminate subheaders row\n",
    "    dataframes.append(df)\n",
    "\n",
    "# Combine all dataframes\n",
    "combined_df = pd.concat(dataframes, ignore_index=True)\n",
    "combined_df['Clinical_info']='YES'\n",
    "print(combined_df.info())  # 309 samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61a321f2-b5e9-4779-b8b2-14b6af800a58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 235 entries, 0 to 234\n",
      "Data columns (total 33 columns):\n",
      " #   Column                           Non-Null Count  Dtype  \n",
      "---  ------                           --------------  -----  \n",
      " 0   ID_CNIO                          235 non-null    object \n",
      " 1   AP number                        204 non-null    object \n",
      " 2   HIST                             205 non-null    object \n",
      " 3   IHC                              232 non-null    object \n",
      " 4   IHC pattern                      232 non-null    object \n",
      " 5   IHC_MMR_class                    232 non-null    object \n",
      " 6   MSI (PCR)                        198 non-null    object \n",
      " 7   Evidence score for MMRd          31 non-null     float64\n",
      " 8   IHC (Diego)                      171 non-null    object \n",
      " 9   PCR (Diego)                      172 non-null    object \n",
      " 10  PASS QC > 2M reads   (Diego)     205 non-null    object \n",
      " 11  FINAL SAMPLES (Diego)            235 non-null    object \n",
      " 12  CLASS ACCORDING IHC/PCR RESULTS  205 non-null    object \n",
      " 13  CLASS ACCORDING TFM  (Diego)     173 non-null    object \n",
      " 14  MSI by MSISensor2 (Diego)        173 non-null    float64\n",
      " 15  TMB                              173 non-null    float64\n",
      " 16  ID2                              170 non-null    float64\n",
      " 17  ID7                              170 non-null    float64\n",
      " 18  SUM ID2+ID7                      173 non-null    float64\n",
      " 19  CNV                              173 non-null    float64\n",
      " 20  % genome altered (Diego)         173 non-null    float64\n",
      " 21  Pol mutation (Diego)             173 non-null    object \n",
      " 22  MMRd mutation (Diego)            173 non-null    object \n",
      " 23  Non-MMRd mutation (Diego)        173 non-null    object \n",
      " 24  TILs counts                      235 non-null    object \n",
      " 25  TILs_CNIO_info                   235 non-null    object \n",
      " 26  TILs_ep_CNIO                     211 non-null    float64\n",
      " 27  TILs_ep_LP                       127 non-null    float64\n",
      " 28  TILs_ep_OTTA                     109 non-null    float64\n",
      " 29  TILs_tu_CNIO                     210 non-null    float64\n",
      " 30  TILs_tu_LP                       88 non-null     float64\n",
      " 31  TILs_raw_ep                      196 non-null    float64\n",
      " 32  TILs_raw_tu                      197 non-null    float64\n",
      "dtypes: float64(15), object(18)\n",
      "memory usage: 60.7+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Add colums with genomic/inmunogenic/IHC/PCR data from Genomic_tils_final_database.xlsx\n",
    "# TILs counts column: YES values give us the information about the number of samples with TILs data and\n",
    "# FINAL SAMPLES (Diego) about the number of samples with genomic information that pass QC filters\n",
    "# Step 1: Read the data\n",
    "df=pd.read_excel('/home/vant/TFM/Final_db/Genomic_tils_final_database.xlsx')\n",
    "# Step 2: See information\n",
    "print(df.info())  # 235 samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "36a3ff77-422f-4950-83a6-8faa4c9c82d9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with same ID_CNIO values in both DataFrames: 209\n",
      "Samples in both DataFrames:\n",
      "['LP1', 'LP2', 'LP3', 'LP4', 'LP5', 'LP6', 'LP7', 'LP8', 'LP9', 'LP10', 'LP11', 'LP12', 'LP13', 'LP14', 'LP15', 'LP16', 'LP17', 'LP18', 'LP19', 'LP20', 'LP21', 'LP22', 'LP24', 'LP25', 'LP26', 'LP27', 'LP28', 'LP29', 'LP30', 'LP31', 'LP32', 'LP33', 'LP34', 'LP35', 'LP36', 'LP37', 'LP38', 'LP39', 'LP40', 'LP41', 'LP42', 'LP43', 'LP44', 'LP45', 'LP46', 'LP47', 'LP48', 'LP49', 'LP50', 'LP51', 'LP53', 'LP55', 'LP56', 'LP57', 'LP58', 'LP59', 'LP60', 'LP61', 'LP62', 'LP63', 'LP64', 'LP65', 'LP66', 'LP67', 'LP68', 'LP69', 'LP70', 'LP71', 'LP72', 'LP73', 'LP74', 'LP75', 'LP76', 'LP77', 'LP78', 'LP79', 'LP80', 'LP81', 'LP82', 'LP83', 'LP84', 'LP85', 'LP86', 'LP87', 'LP88', 'LP89', 'LP90', 'LP91', 'LP92', 'LP93', 'LP94', 'LP95', 'LP96', 'LP97', 'LP98', 'LP99', 'LP100', 'LP102', 'LP103', 'LP104', 'LP105', 'LP106', 'LP109', 'LP110', 'LP114', 'LP115', 'LP116', 'LP117', 'LP119', 'LP121', 'MDA1', 'MDA2', 'MDA3', 'MDA4', 'MDA5', 'MDA6', 'MDA7', 'MDA8', 'MDA9', 'MDA10', 'MDA11', 'MDA12', 'MDA13', 'MDA14', 'MDA15', 'MDA16', 'MDA17', 'MDA18', 'MDA19', 'MDA20', 'MDA21', 'MDA22', 'MDA23', 'MDA24', 'MDA25', 'MDA26', 'MDA27', 'MDA28', 'MDA29', 'MDA30', 'MDA31', 'MDA32', 'MDA33', 'MDA34', 'MDA35', 'MDA36', 'MDA37', 'MDA38', 'MDA39', 'MDA40', 'MDA41', 'MDA42', 'MDA43', 'MDA44', 'MDA45', 'MDA46', 'MDA47', 'MDA48', 'MDA49', 'MDA50', 'MDA51_T1_CCC', 'MDA51_T2_EC', 'OVE4', 'OVE6', 'OVE10', 'OVE14', 'OVE16', 'OVE17', 'OVE27', 'OVE28', 'OVE31', 'OVE34', 'OVE35', 'OVE37', 'OVE38', 'OVE41', 'OVE42', 'OVE43', 'OVE44', 'OVE49', 'OVE57', 'OVE58', 'OVE59', 'OVE61', 'OVE63', 'OVE64', 'OVE72', 'OVE74', 'OVE76', 'OVE77', 'OVE80', 'OVE83', 'OVE84', 'OVE85', 'OVE86', 'OVE90', 'OVE92', 'OVE93', 'OVE94', 'OVE96', 'OVE104', 'OVE108', 'RVB1', 'RVB2', 'RVB3', 'RVB4', 'RVB5', 'RVB6', 'RVB7']\n",
      "Number of samples only in combined_df: 100\n",
      "Samples only in combined_df-clinical data:\n",
      "['LP23', 'LP101', 'LP107', 'LP108', 'LP111', 'LP112', 'LP113', 'LP118', 'LP120', 'OVE1', 'OVE2', '08T402 / OVE3', '08T405 / OVE5', 'OVE7', 'OVE8', 'OVE9', 'OVE11', '08T395 / OVE12', '08T398 / OVE13', 'OVE15', '08T406 / OVE18', '08T396 / OVE19', 'OVE20', 'OVE21', 'OVE22', 'OVE23', '08T399 / OVE24', 'OVE25', 'OVE26', '08T400 / OVE29', 'OVE30', 'OVE32', 'OVE33', 'OVE36', '08T397 / OVE39', '08T407 / OVE40', '08T394 / OVE45', 'OVE46', 'OVE47', '08T403 / OVE48', '08T401 / OVE50', '08T404 / OVE51', 'OVE52', 'OVE53', 'OVE54', 'OVE55', 'OVE56', 'OVE60', 'OVE62', 'OVE65', 'OVE66', 'OVE67', 'OVE68', 'OVE69', 'OVE70', 'OVE71', 'OVE73', 'OVE75', 'OVE78', 'OVE79', 'OVE81', 'OVE82', 'OVE87', 'OVE88', 'OVE89', 'OVE91', 'OVE95', 'OVE97', 'OVE98', 'OVE99', 'OVE100', 'OVE101', 'OVE102', 'OVE103', 'OVE105', 'OVE106', 'OVE107', 'OVE109', 'OVE110', 'OVE111', 'OVE112', 'OVE113', 'OVE114', 'OVE115', 'OVE116', 'OVE117', 'OVE118', 'OVE119', 'OVE120', 'OVE121', 'OVE122', 'OVE123', 'OVE124', 'OVE125', 'OVE126', 'OVE127', 'OVE128', 'OVE129', 'RVB8', 'RVB9']\n",
      "Number of samples only in df: 26\n",
      "Samples only in df-genomic/tils/IHC/PCR:\n",
      "['LP52', 'LP54', 'LP124', '03B07829', '07B23784', '10B36563', '11B23809', '11B5978', '07T25', '08T96', '09T134', '07T29', '08T107', '09T-114', '09T-117-T', '09T-119', '09T45 ', 'OVE13', 'OVE19', 'OVE24', 'OVE39', 'OVE48', '08T94', '09T237', '11T82', '16T131']\n"
     ]
    }
   ],
   "source": [
    "# Eliminate additional spaces in the column names\n",
    "df.columns = df.columns.str.strip()\n",
    "combined_df.columns = combined_df.columns.str.strip()\n",
    "\n",
    "# Step 1: Find ID_CNIO values present in both DataFrames\n",
    "common_ids = pd.merge(combined_df[['ID_CNIO']], df[['ID_CNIO']], on='ID_CNIO', how='inner')['ID_CNIO']\n",
    "\n",
    "# Step 2: Find unique ID_CNIO values in combined_df (clinical database) , but not in df (genomic+gold standard+tils database)\n",
    "only_in_combined_df = combined_df[~combined_df['ID_CNIO'].isin(df['ID_CNIO'])]['ID_CNIO']\n",
    "\n",
    "# Step 3: Find unique ID_CNIO values in df (genomic+GS+tils database), but not in combined_df (clinical database)\n",
    "only_in_df = df[~df['ID_CNIO'].isin(combined_df['ID_CNIO'])]['ID_CNIO']\n",
    "\n",
    "# Step 4: Show the results\n",
    "print(f\"Number of samples with same ID_CNIO values in both DataFrames: {len(common_ids)}\")\n",
    "print(f\"Samples in both DataFrames:\\n{common_ids.to_list()}\")\n",
    "\n",
    "print(f\"Number of samples only in combined_df: {len(only_in_combined_df)}\")\n",
    "print(f\"Samples only in combined_df-clinical data:\\n{only_in_combined_df.to_list()}\")\n",
    "\n",
    "print(f\"Number of samples only in df: {len(only_in_df)}\")\n",
    "print(f\"Samples only in df-genomic/tils/IHC/PCR:\\n{only_in_df.to_list()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7dd23ba4-1803-40f4-8887-1cf35547aaf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total samples with the same ID_CNIO in both DataFrames: 214\n",
      "Samples in both DataFrames:\n",
      "['LP1', 'LP2', 'LP3', 'LP4', 'LP5', 'LP6', 'LP7', 'LP8', 'LP9', 'LP10', 'LP11', 'LP12', 'LP13', 'LP14', 'LP15', 'LP16', 'LP17', 'LP18', 'LP19', 'LP20', 'LP21', 'LP22', 'LP24', 'LP25', 'LP26', 'LP27', 'LP28', 'LP29', 'LP30', 'LP31', 'LP32', 'LP33', 'LP34', 'LP35', 'LP36', 'LP37', 'LP38', 'LP39', 'LP40', 'LP41', 'LP42', 'LP43', 'LP44', 'LP45', 'LP46', 'LP47', 'LP48', 'LP49', 'LP50', 'LP51', 'LP53', 'LP55', 'LP56', 'LP57', 'LP58', 'LP59', 'LP60', 'LP61', 'LP62', 'LP63', 'LP64', 'LP65', 'LP66', 'LP67', 'LP68', 'LP69', 'LP70', 'LP71', 'LP72', 'LP73', 'LP74', 'LP75', 'LP76', 'LP77', 'LP78', 'LP79', 'LP80', 'LP81', 'LP82', 'LP83', 'LP84', 'LP85', 'LP86', 'LP87', 'LP88', 'LP89', 'LP90', 'LP91', 'LP92', 'LP93', 'LP94', 'LP95', 'LP96', 'LP97', 'LP98', 'LP99', 'LP100', 'LP102', 'LP103', 'LP104', 'LP105', 'LP106', 'LP109', 'LP110', 'LP114', 'LP115', 'LP116', 'LP117', 'LP119', 'LP121', 'MDA1', 'MDA2', 'MDA3', 'MDA4', 'MDA5', 'MDA6', 'MDA7', 'MDA8', 'MDA9', 'MDA10', 'MDA11', 'MDA12', 'MDA13', 'MDA14', 'MDA15', 'MDA16', 'MDA17', 'MDA18', 'MDA19', 'MDA20', 'MDA21', 'MDA22', 'MDA23', 'MDA24', 'MDA25', 'MDA26', 'MDA27', 'MDA28', 'MDA29', 'MDA30', 'MDA31', 'MDA32', 'MDA33', 'MDA34', 'MDA35', 'MDA36', 'MDA37', 'MDA38', 'MDA39', 'MDA40', 'MDA41', 'MDA42', 'MDA43', 'MDA44', 'MDA45', 'MDA46', 'MDA47', 'MDA48', 'MDA49', 'MDA50', 'MDA51_T1_CCC', 'MDA51_T2_EC', 'OVE4', 'OVE6', 'OVE10', 'OVE13', 'OVE14', 'OVE16', 'OVE17', 'OVE19', 'OVE24', 'OVE27', 'OVE28', 'OVE31', 'OVE34', 'OVE35', 'OVE37', 'OVE38', 'OVE39', 'OVE41', 'OVE42', 'OVE43', 'OVE44', 'OVE48', 'OVE49', 'OVE57', 'OVE58', 'OVE59', 'OVE61', 'OVE63', 'OVE64', 'OVE72', 'OVE74', 'OVE76', 'OVE77', 'OVE80', 'OVE83', 'OVE84', 'OVE85', 'OVE86', 'OVE90', 'OVE92', 'OVE93', 'OVE94', 'OVE96', 'OVE104', 'OVE108', 'RVB1', 'RVB2', 'RVB3', 'RVB4', 'RVB5', 'RVB6', 'RVB7']\n",
      "Total samples only in combined_df: 95\n",
      "Samples only in combined_df-clinical data:\n",
      "['LP23', 'LP101', 'LP107', 'LP108', 'LP111', 'LP112', 'LP113', 'LP118', 'LP120', 'OVE1', 'OVE2', 'OVE3', 'OVE5', 'OVE7', 'OVE8', 'OVE9', 'OVE11', 'OVE12', 'OVE15', 'OVE18', 'OVE20', 'OVE21', 'OVE22', 'OVE23', 'OVE25', 'OVE26', 'OVE29', 'OVE30', 'OVE32', 'OVE33', 'OVE36', 'OVE40', 'OVE45', 'OVE46', 'OVE47', 'OVE50', 'OVE51', 'OVE52', 'OVE53', 'OVE54', 'OVE55', 'OVE56', 'OVE60', 'OVE62', 'OVE65', 'OVE66', 'OVE67', 'OVE68', 'OVE69', 'OVE70', 'OVE71', 'OVE73', 'OVE75', 'OVE78', 'OVE79', 'OVE81', 'OVE82', 'OVE87', 'OVE88', 'OVE89', 'OVE91', 'OVE95', 'OVE97', 'OVE98', 'OVE99', 'OVE100', 'OVE101', 'OVE102', 'OVE103', 'OVE105', 'OVE106', 'OVE107', 'OVE109', 'OVE110', 'OVE111', 'OVE112', 'OVE113', 'OVE114', 'OVE115', 'OVE116', 'OVE117', 'OVE118', 'OVE119', 'OVE120', 'OVE121', 'OVE122', 'OVE123', 'OVE124', 'OVE125', 'OVE126', 'OVE127', 'OVE128', 'OVE129', 'RVB8', 'RVB9']\n",
      "Total samples only in df: 21\n",
      "Samples only in df-genomic/tils/IHC/PCR:\n",
      "['LP52', 'LP54', 'LP124', '03B07829', '07B23784', '10B36563', '11B23809', '11B5978', '07T25', '08T96', '09T134', '07T29', '08T107', '09T-114', '09T-117-T', '09T-119', '09T45', '08T94', '09T237', '11T82', '16T131']\n"
     ]
    }
   ],
   "source": [
    "# After the last operation, I observe that some ID_CNIO values in combined_df, such as 08T399 / OVE24, for example,\n",
    "# do not match with OVE24 in ID_CNIO in df. To solve this:\n",
    "# Extract the relevant part of ID_CNIO in combined_df\n",
    "import re\n",
    "\n",
    "# Remove irrelevant prefixes from IDs in combined_df, keeping the suffixes. In this way, I replace any digit followed by\n",
    "# a 'T' and then any digit one or more times, optionally followed by zero or more spaces, and replace it with an empty string.\n",
    "combined_df['ID_CNIO_corrected'] = combined_df['ID_CNIO'].apply(lambda x: re.sub(r'^\\d+T\\d+\\s*/\\s*', '', x))\n",
    "# The IDs in df are already correct and do not require modification.\n",
    "df['ID_CNIO_corrected'] = df['ID_CNIO'].str.strip()\n",
    "# Step 3: Find the ID_CNIO present in both DataFrames\n",
    "common_ids = pd.merge(combined_df[['ID_CNIO_corrected']], df[['ID_CNIO_corrected']], on='ID_CNIO_corrected', how='inner')['ID_CNIO_corrected']\n",
    "\n",
    "# Step 4: Find the unique ID_CNIO in combined_df but not in df\n",
    "only_in_combined_df = combined_df[~combined_df['ID_CNIO_corrected'].isin(df['ID_CNIO_corrected'])]['ID_CNIO_corrected']\n",
    "\n",
    "# Step 5: Find the unique ID_CNIO in df but not in combined_df\n",
    "only_in_df = df[~df['ID_CNIO_corrected'].isin(combined_df['ID_CNIO_corrected'])]['ID_CNIO_corrected']\n",
    "\n",
    "# Step 6: Display the results\n",
    "print(f\"Total samples with the same ID_CNIO in both DataFrames: {len(common_ids)}\")\n",
    "print(f\"Samples in both DataFrames:\\n{common_ids.to_list()}\")\n",
    "\n",
    "print(f\"Total samples only in combined_df: {len(only_in_combined_df)}\")\n",
    "print(f\"Samples only in combined_df-clinical data:\\n{only_in_combined_df.to_list()}\")\n",
    "\n",
    "print(f\"Total samples only in df: {len(only_in_df)}\")\n",
    "print(f\"Samples only in df-genomic/tils/IHC/PCR:\\n{only_in_df.to_list()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c2f64bdc-b01a-4311-aa51-9e76d50da4df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 309 entries, 0 to 308\n",
      "Data columns (total 27 columns):\n",
      " #   Column                    Non-Null Count  Dtype \n",
      "---  ------                    --------------  ----- \n",
      " 0   ID_CNIO                   309 non-null    object\n",
      " 1   HISTOLOGY                 306 non-null    object\n",
      " 2   AGE                       309 non-null    object\n",
      " 3   GRADE                     243 non-null    object\n",
      " 4   FIGO                      275 non-null    object\n",
      " 5   FIGOL                     276 non-null    object\n",
      " 6   FIGOa                     275 non-null    object\n",
      " 7   NEOADJUVANT_TREATMENT     233 non-null    object\n",
      " 8   ADJUVANT_TREATMENT        279 non-null    object\n",
      " 9   TYPE_ADJUVANT             197 non-null    object\n",
      " 10  RESIDUAL                  224 non-null    object\n",
      " 11  RESIDUALa                 224 non-null    object\n",
      " 12  OS_CNIO                   284 non-null    object\n",
      " 13  VITAL STATUS              289 non-null    object\n",
      " 14  MMR STATUS                179 non-null    object\n",
      " 15  BRCA STATUS               169 non-null    object\n",
      " 16  FAMILIAL                  266 non-null    object\n",
      " 17  DATE_OS                   308 non-null    object\n",
      " 18  ID_ORIGINAL               299 non-null    object\n",
      " 19  ID_ORIGINAL_NHC_BBANK     283 non-null    object\n",
      " 20  OTHER_ADJ_TREAT           14 non-null     object\n",
      " 21  PARTIAL DATE DEATH_LASTv  61 non-null     object\n",
      " 22  PARTIAL DATE DIAGNOSIS    61 non-null     object\n",
      " 23  HISTOLOGY_DETAIL          129 non-null    object\n",
      " 24  DATE_AGE                  129 non-null    object\n",
      " 25  Clinical_info             309 non-null    object\n",
      " 26  ID_CNIO_corrected         309 non-null    object\n",
      "dtypes: object(27)\n",
      "memory usage: 65.3+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 235 entries, 0 to 234\n",
      "Data columns (total 34 columns):\n",
      " #   Column                           Non-Null Count  Dtype  \n",
      "---  ------                           --------------  -----  \n",
      " 0   ID_CNIO                          235 non-null    object \n",
      " 1   AP number                        204 non-null    object \n",
      " 2   HIST                             205 non-null    object \n",
      " 3   IHC                              232 non-null    object \n",
      " 4   IHC pattern                      232 non-null    object \n",
      " 5   IHC_MMR_class                    232 non-null    object \n",
      " 6   MSI (PCR)                        198 non-null    object \n",
      " 7   Evidence score for MMRd          31 non-null     float64\n",
      " 8   IHC (Diego)                      171 non-null    object \n",
      " 9   PCR (Diego)                      172 non-null    object \n",
      " 10  PASS QC > 2M reads   (Diego)     205 non-null    object \n",
      " 11  FINAL SAMPLES (Diego)            235 non-null    object \n",
      " 12  CLASS ACCORDING IHC/PCR RESULTS  205 non-null    object \n",
      " 13  CLASS ACCORDING TFM  (Diego)     173 non-null    object \n",
      " 14  MSI by MSISensor2 (Diego)        173 non-null    float64\n",
      " 15  TMB                              173 non-null    float64\n",
      " 16  ID2                              170 non-null    float64\n",
      " 17  ID7                              170 non-null    float64\n",
      " 18  SUM ID2+ID7                      173 non-null    float64\n",
      " 19  CNV                              173 non-null    float64\n",
      " 20  % genome altered (Diego)         173 non-null    float64\n",
      " 21  Pol mutation (Diego)             173 non-null    object \n",
      " 22  MMRd mutation (Diego)            173 non-null    object \n",
      " 23  Non-MMRd mutation (Diego)        173 non-null    object \n",
      " 24  TILs counts                      235 non-null    object \n",
      " 25  TILs_CNIO_info                   235 non-null    object \n",
      " 26  TILs_ep_CNIO                     211 non-null    float64\n",
      " 27  TILs_ep_LP                       127 non-null    float64\n",
      " 28  TILs_ep_OTTA                     109 non-null    float64\n",
      " 29  TILs_tu_CNIO                     210 non-null    float64\n",
      " 30  TILs_tu_LP                       88 non-null     float64\n",
      " 31  TILs_raw_ep                      196 non-null    float64\n",
      " 32  TILs_raw_tu                      197 non-null    float64\n",
      " 33  ID_CNIO_corrected                235 non-null    object \n",
      "dtypes: float64(15), object(19)\n",
      "memory usage: 62.5+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(combined_df.info())\n",
    "print(df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f4221cf-1330-4c54-a35e-1329bfa22cac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  ID_CNIO HISTOLOGY AGE GRADE FIGO FIGOL FIGOa NEOADJUVANT_TREATMENT  \\\n",
      "0     LP1         0  50     2    3     0     1                     0   \n",
      "1     LP2         0  42     2    3     0     1                     0   \n",
      "2     LP3         0  68     2    5     0     2                     0   \n",
      "3     LP4         0  50     1    6     0     2                     0   \n",
      "4     LP5         0  82     1    1     0     1                     0   \n",
      "\n",
      "  ADJUVANT_TREATMENT TYPE_ADJUVANT  ... Non-MMRd mutation (Diego) TILs counts  \\\n",
      "0                  1             0  ...                        NO         YES   \n",
      "1                  1             0  ...                        NO         YES   \n",
      "2                  1             0  ...                        NO         YES   \n",
      "3                  1             0  ...                        NO         YES   \n",
      "4                  0           NaN  ...                        NO         YES   \n",
      "\n",
      "  TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP TILs_ep_OTTA TILs_tu_CNIO TILs_tu_LP  \\\n",
      "0   YES_complete          1.0        9.0          9.0          2.0        2.0   \n",
      "1   YES_complete          2.0        0.0          0.0          2.0        0.0   \n",
      "2   YES_complete          3.0        3.0          1.0          3.0        3.0   \n",
      "3   YES_complete          2.0        3.0          2.0          2.0        3.0   \n",
      "4             NO          NaN        1.0          0.0          NaN        1.0   \n",
      "\n",
      "  TILs_raw_ep TILs_raw_tu  \n",
      "0         2.0         6.0  \n",
      "1         4.0         7.0  \n",
      "2        56.0        78.0  \n",
      "3         5.0        15.0  \n",
      "4         NaN         NaN  \n",
      "\n",
      "[5 rows x 57 columns]\n",
      "Length of the complete df is: 330\n"
     ]
    }
   ],
   "source": [
    "# I eliminate every space in the name of the columns. There are samples in df with no information about the clinics (combined_df) and viceversa\n",
    "df.columns = df.columns.str.strip()\n",
    "combined_df.columns = combined_df.columns.str.strip()\n",
    "# List of specific columns in df I want to add to the new dataframe \n",
    "columns_to_merge = [\n",
    "    'IHC', 'IHC pattern','IHC_MMR_class','MSI (PCR)', 'Evidence score for MMRd','IHC (Diego)', 'PCR (Diego)', 'PASS QC > 2M reads   (Diego)', \n",
    "    'FINAL SAMPLES (Diego)', 'CLASS ACCORDING IHC/PCR RESULTS', 'CLASS ACCORDING TFM  (Diego)', \n",
    "    'MSI by MSISensor2 (Diego)', 'TMB', 'ID2', 'ID7', 'SUM ID2+ID7', 'CNV', \n",
    "    '% genome altered (Diego)', 'Pol mutation (Diego)', 'MMRd mutation (Diego)', \n",
    "    'Non-MMRd mutation (Diego)', 'TILs counts', 'TILs_CNIO_info','TILs_ep_CNIO', 'TILs_ep_LP', \n",
    "    'TILs_ep_OTTA', 'TILs_tu_CNIO', 'TILs_tu_LP','TILs_raw_ep','TILs_raw_tu'\n",
    "]\n",
    "# Combine DataFrames keeping all rows\n",
    "final_df = pd.merge(\n",
    "    combined_df,\n",
    "    df[['ID_CNIO_corrected'] + columns_to_merge],  # Select only the necessary columns in df\n",
    "    on='ID_CNIO_corrected',\n",
    "    how='outer'  # 'outer' allows to keep all rows in combined_df and df\n",
    ")\n",
    "\n",
    "# Show the final DataFrame \n",
    "print(final_df.head())\n",
    "print(\"Length of the complete df is:\",len(final_df))\n",
    "final_df.to_excel(\"/home/vant/TFM/Final_db/Combined_df_clinical_genomic_tils.xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4531d429-7021-448d-b482-f0bea1c45c4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 330 entries, 0 to 329\n",
      "Data columns (total 57 columns):\n",
      " #   Column                           Non-Null Count  Dtype  \n",
      "---  ------                           --------------  -----  \n",
      " 0   ID_CNIO                          309 non-null    object \n",
      " 1   HISTOLOGY                        306 non-null    object \n",
      " 2   AGE                              309 non-null    object \n",
      " 3   GRADE                            243 non-null    object \n",
      " 4   FIGO                             275 non-null    object \n",
      " 5   FIGOL                            276 non-null    object \n",
      " 6   FIGOa                            275 non-null    object \n",
      " 7   NEOADJUVANT_TREATMENT            233 non-null    object \n",
      " 8   ADJUVANT_TREATMENT               279 non-null    object \n",
      " 9   TYPE_ADJUVANT                    197 non-null    object \n",
      " 10  RESIDUAL                         224 non-null    object \n",
      " 11  RESIDUALa                        224 non-null    object \n",
      " 12  OS_CNIO                          284 non-null    object \n",
      " 13  VITAL STATUS                     289 non-null    object \n",
      " 14  MMR STATUS                       179 non-null    object \n",
      " 15  BRCA STATUS                      169 non-null    object \n",
      " 16  FAMILIAL                         266 non-null    object \n",
      " 17  DATE_OS                          308 non-null    object \n",
      " 18  ID_ORIGINAL                      299 non-null    object \n",
      " 19  ID_ORIGINAL_NHC_BBANK            283 non-null    object \n",
      " 20  OTHER_ADJ_TREAT                  14 non-null     object \n",
      " 21  PARTIAL DATE DEATH_LASTv         61 non-null     object \n",
      " 22  PARTIAL DATE DIAGNOSIS           61 non-null     object \n",
      " 23  HISTOLOGY_DETAIL                 129 non-null    object \n",
      " 24  DATE_AGE                         129 non-null    object \n",
      " 25  Clinical_info                    309 non-null    object \n",
      " 26  ID_CNIO_corrected                330 non-null    object \n",
      " 27  IHC                              232 non-null    object \n",
      " 28  IHC pattern                      232 non-null    object \n",
      " 29  IHC_MMR_class                    232 non-null    object \n",
      " 30  MSI (PCR)                        198 non-null    object \n",
      " 31  Evidence score for MMRd          31 non-null     float64\n",
      " 32  IHC (Diego)                      171 non-null    object \n",
      " 33  PCR (Diego)                      172 non-null    object \n",
      " 34  PASS QC > 2M reads   (Diego)     205 non-null    object \n",
      " 35  FINAL SAMPLES (Diego)            235 non-null    object \n",
      " 36  CLASS ACCORDING IHC/PCR RESULTS  205 non-null    object \n",
      " 37  CLASS ACCORDING TFM  (Diego)     173 non-null    object \n",
      " 38  MSI by MSISensor2 (Diego)        173 non-null    float64\n",
      " 39  TMB                              173 non-null    float64\n",
      " 40  ID2                              170 non-null    float64\n",
      " 41  ID7                              170 non-null    float64\n",
      " 42  SUM ID2+ID7                      173 non-null    float64\n",
      " 43  CNV                              173 non-null    float64\n",
      " 44  % genome altered (Diego)         173 non-null    float64\n",
      " 45  Pol mutation (Diego)             173 non-null    object \n",
      " 46  MMRd mutation (Diego)            173 non-null    object \n",
      " 47  Non-MMRd mutation (Diego)        173 non-null    object \n",
      " 48  TILs counts                      235 non-null    object \n",
      " 49  TILs_CNIO_info                   235 non-null    object \n",
      " 50  TILs_ep_CNIO                     211 non-null    float64\n",
      " 51  TILs_ep_LP                       127 non-null    float64\n",
      " 52  TILs_ep_OTTA                     109 non-null    float64\n",
      " 53  TILs_tu_CNIO                     210 non-null    float64\n",
      " 54  TILs_tu_LP                       88 non-null     float64\n",
      " 55  TILs_raw_ep                      196 non-null    float64\n",
      " 56  TILs_raw_tu                      197 non-null    float64\n",
      "dtypes: float64(15), object(42)\n",
      "memory usage: 147.1+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(final_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6d07b1c0-2b56-458a-b23a-31a10b93c907",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 330 entries, 0 to 329\n",
      "Data columns (total 56 columns):\n",
      " #   Column                     Non-Null Count  Dtype  \n",
      "---  ------                     --------------  -----  \n",
      " 0   HISTOLOGY                  306 non-null    object \n",
      " 1   AGE                        309 non-null    object \n",
      " 2   GRADE                      243 non-null    object \n",
      " 3   FIGO                       275 non-null    object \n",
      " 4   FIGOL                      276 non-null    object \n",
      " 5   FIGOa                      275 non-null    object \n",
      " 6   NEOADJUVANT_TREATMENT      233 non-null    object \n",
      " 7   ADJUVANT_TREATMENT         279 non-null    object \n",
      " 8   TYPE_ADJUVANT              197 non-null    object \n",
      " 9   RESIDUAL                   224 non-null    object \n",
      " 10  RESIDUALa                  224 non-null    object \n",
      " 11  OS_CNIO                    284 non-null    object \n",
      " 12  VITAL STATUS               289 non-null    object \n",
      " 13  MMR STATUS                 179 non-null    object \n",
      " 14  BRCA STATUS                169 non-null    object \n",
      " 15  FAMILIAL                   266 non-null    object \n",
      " 16  DATE_OS                    308 non-null    object \n",
      " 17  ID_ORIGINAL                299 non-null    object \n",
      " 18  ID_ORIGINAL_NHC_BBANK      283 non-null    object \n",
      " 19  OTHER_ADJ_TREAT            14 non-null     object \n",
      " 20  PARTIAL DATE DEATH_LASTv   61 non-null     object \n",
      " 21  PARTIAL DATE DIAGNOSIS     61 non-null     object \n",
      " 22  HISTOLOGY_DETAIL           129 non-null    object \n",
      " 23  DATE_AGE                   129 non-null    object \n",
      " 24  Clinical_info              309 non-null    object \n",
      " 25  ID_CNIO                    330 non-null    object \n",
      " 26  IHC_defective_MMR_protein  232 non-null    object \n",
      " 27  IHC pattern                232 non-null    object \n",
      " 28  IHC_MMR_class              232 non-null    object \n",
      " 29  PCR                        198 non-null    object \n",
      " 30  Evidence score for MMRd    31 non-null     float64\n",
      " 31  IHCd                       171 non-null    object \n",
      " 32  PCRd                       172 non-null    object \n",
      " 33  QC_filters                 205 non-null    object \n",
      " 34  Genomic_info               235 non-null    object \n",
      " 35  MMR_final_status           205 non-null    object \n",
      " 36  CLASSd                     173 non-null    object \n",
      " 37  MSI_sensor2                173 non-null    float64\n",
      " 38  TMB                        173 non-null    float64\n",
      " 39  ID2                        170 non-null    float64\n",
      " 40  ID7                        170 non-null    float64\n",
      " 41  SUM ID2+ID7                173 non-null    float64\n",
      " 42  CNV                        173 non-null    float64\n",
      " 43  %genome_altered            173 non-null    float64\n",
      " 44  Pol_mutation               173 non-null    object \n",
      " 45  MMRd_mutation              173 non-null    object \n",
      " 46  Non-MMRd_mutation          173 non-null    object \n",
      " 47  TILs_score_info            235 non-null    object \n",
      " 48  TILs_CNIO_info             235 non-null    object \n",
      " 49  TILs_ep_CNIO               211 non-null    float64\n",
      " 50  TILs_ep_LP                 127 non-null    float64\n",
      " 51  TILs_ep_OTTA               109 non-null    float64\n",
      " 52  TILs_tu_CNIO               210 non-null    float64\n",
      " 53  TILs_tu_LP                 88 non-null     float64\n",
      " 54  TILs_raw_ep                196 non-null    float64\n",
      " 55  TILs_raw_tu                197 non-null    float64\n",
      "dtypes: float64(15), object(41)\n",
      "memory usage: 144.5+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Eliminate ID_CNIO because it is repeated, rename ID_CNIO_corrected to ID_CNIO\n",
    "final_df.drop(columns=['ID_CNIO'],inplace=True)\n",
    "# Create a dictitonary to rename the columns names.\n",
    "final_df.rename(columns={\n",
    "    'ID_CNIO_corrected': 'ID_CNIO',\n",
    "    'MSI (PCR)' :'PCR',\n",
    "    'IHC': 'IHC_defective_MMR_protein',\n",
    "    'IHC (Diego)':'IHCd',\n",
    "    'PCR (Diego)':'PCRd',\n",
    "    'PASS QC > 2M reads   (Diego)':'QC_filters',\n",
    "    'FINAL SAMPLES (Diego)':'Genomic_info', # 173.Remember Diego gave us data of 4 extra samples that didn´t appear in its previous database version\n",
    "    'CLASS ACCORDING IHC/PCR RESULTS':'MMR_final_status', \n",
    "    'CLASS ACCORDING TFM  (Diego)':'CLASSd',\n",
    "    'MSI by MSISensor2 (Diego)':'MSI_sensor2',\n",
    "     '% genome altered (Diego)':'%genome_altered',\n",
    "    'Pol mutation (Diego)':'Pol_mutation',\n",
    "    'MMRd mutation (Diego)':'MMRd_mutation',\n",
    "    'Non-MMRd mutation (Diego)':'Non-MMRd_mutation',\n",
    "    'TILs counts':'TILs_score_info',\n",
    "    \n",
    "}, inplace=True)\n",
    "\n",
    "# Check the changes\n",
    "print(final_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6a27c4c6-3900-42fe-a99d-e5fe9099c247",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 330 entries, 0 to 329\n",
      "Data columns (total 56 columns):\n",
      " #   Column                     Non-Null Count  Dtype  \n",
      "---  ------                     --------------  -----  \n",
      " 0   ID_CNIO                    330 non-null    object \n",
      " 1   ID_ORIGINAL                299 non-null    object \n",
      " 2   ID_ORIGINAL_NHC_BBANK      283 non-null    object \n",
      " 3   HISTOLOGY                  306 non-null    object \n",
      " 4   HISTOLOGY_DETAIL           129 non-null    object \n",
      " 5   AGE                        309 non-null    object \n",
      " 6   GRADE                      243 non-null    object \n",
      " 7   FIGO                       275 non-null    object \n",
      " 8   FIGOL                      276 non-null    object \n",
      " 9   FIGOa                      275 non-null    object \n",
      " 10  NEOADJUVANT_TREATMENT      233 non-null    object \n",
      " 11  ADJUVANT_TREATMENT         279 non-null    object \n",
      " 12  TYPE_ADJUVANT              197 non-null    object \n",
      " 13  OTHER_ADJ_TREAT            14 non-null     object \n",
      " 14  RESIDUAL                   224 non-null    object \n",
      " 15  RESIDUALa                  224 non-null    object \n",
      " 16  OS_CNIO                    284 non-null    object \n",
      " 17  VITAL STATUS               289 non-null    object \n",
      " 18  MMR STATUS                 179 non-null    object \n",
      " 19  BRCA STATUS                169 non-null    object \n",
      " 20  FAMILIAL                   266 non-null    object \n",
      " 21  DATE_AGE                   129 non-null    object \n",
      " 22  DATE_OS                    308 non-null    object \n",
      " 23  PARTIAL DATE DIAGNOSIS     61 non-null     object \n",
      " 24  PARTIAL DATE DEATH_LASTv   61 non-null     object \n",
      " 25  Clinical_info              309 non-null    object \n",
      " 26  IHC_defective_MMR_protein  232 non-null    object \n",
      " 27  IHC pattern                232 non-null    object \n",
      " 28  IHC_MMR_class              232 non-null    object \n",
      " 29  PCR                        198 non-null    object \n",
      " 30  Evidence score for MMRd    31 non-null     float64\n",
      " 31  IHCd                       171 non-null    object \n",
      " 32  PCRd                       172 non-null    object \n",
      " 33  QC_filters                 205 non-null    object \n",
      " 34  Genomic_info               235 non-null    object \n",
      " 35  MMR_final_status           205 non-null    object \n",
      " 36  CLASSd                     173 non-null    object \n",
      " 37  MSI_sensor2                173 non-null    float64\n",
      " 38  TMB                        173 non-null    float64\n",
      " 39  ID2                        170 non-null    float64\n",
      " 40  ID7                        170 non-null    float64\n",
      " 41  SUM ID2+ID7                173 non-null    float64\n",
      " 42  CNV                        173 non-null    float64\n",
      " 43  %genome_altered            173 non-null    float64\n",
      " 44  Pol_mutation               173 non-null    object \n",
      " 45  MMRd_mutation              173 non-null    object \n",
      " 46  Non-MMRd_mutation          173 non-null    object \n",
      " 47  TILs_score_info            235 non-null    object \n",
      " 48  TILs_CNIO_info             235 non-null    object \n",
      " 49  TILs_ep_CNIO               211 non-null    float64\n",
      " 50  TILs_ep_LP                 127 non-null    float64\n",
      " 51  TILs_ep_OTTA               109 non-null    float64\n",
      " 52  TILs_tu_CNIO               210 non-null    float64\n",
      " 53  TILs_tu_LP                 88 non-null     float64\n",
      " 54  TILs_raw_ep                196 non-null    float64\n",
      " 55  TILs_raw_tu                197 non-null    float64\n",
      "dtypes: float64(15), object(41)\n",
      "memory usage: 144.5+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# New order of the columns\n",
    "new_order = [\n",
    "    'ID_CNIO', 'ID_ORIGINAL','ID_ORIGINAL_NHC_BBANK','HISTOLOGY', 'HISTOLOGY_DETAIL','AGE', 'GRADE', 'FIGO',\n",
    "    'FIGOL', 'FIGOa', 'NEOADJUVANT_TREATMENT', 'ADJUVANT_TREATMENT', 'TYPE_ADJUVANT', 'OTHER_ADJ_TREAT',\n",
    "    'RESIDUAL', 'RESIDUALa', 'OS_CNIO', 'VITAL STATUS', 'MMR STATUS', 'BRCA STATUS','FAMILIAL','DATE_AGE',\n",
    "    'DATE_OS','PARTIAL DATE DIAGNOSIS','PARTIAL DATE DEATH_LASTv','Clinical_info','IHC_defective_MMR_protein', 'IHC pattern','IHC_MMR_class',\n",
    "    'PCR','Evidence score for MMRd','IHCd','PCRd','QC_filters','Genomic_info','MMR_final_status','CLASSd','MSI_sensor2','TMB','ID2','ID7','SUM ID2+ID7','CNV','%genome_altered','Pol_mutation','MMRd_mutation',\n",
    "    'Non-MMRd_mutation','TILs_score_info', 'TILs_CNIO_info','TILs_ep_CNIO','TILs_ep_LP','TILs_ep_OTTA','TILs_tu_CNIO','TILs_tu_LP','TILs_raw_ep',\n",
    "    'TILs_raw_tu'\n",
    "]\n",
    "final_df=final_df[new_order]\n",
    "print(final_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "36ebe7ef-51eb-4cda-9c56-e6b19ba1f266",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 235 entries, 0 to 234\n",
      "Data columns (total 5 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   ID CNIO                235 non-null    object\n",
      " 1   ORIGINAL ID_AP         226 non-null    object\n",
      " 2   ORIGINAL ID_NHC_BBANK  204 non-null    object\n",
      " 3   SUBTIPO HISTOLÓGICO    235 non-null    object\n",
      " 4   SERIE                  235 non-null    object\n",
      "dtypes: object(5)\n",
      "memory usage: 9.3+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 235 entries, 0 to 234\n",
      "Data columns (total 5 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   ID CNIO                235 non-null    object\n",
      " 1   ORIGINAL ID_AP         226 non-null    object\n",
      " 2   ORIGINAL ID_NHC_BBANK  204 non-null    object\n",
      " 3   SUBTIPO HISTOLÓGICO    235 non-null    object\n",
      " 4   SERIE                  235 non-null    object\n",
      "dtypes: object(5)\n",
      "memory usage: 9.3+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 235 entries, 0 to 234\n",
      "Data columns (total 5 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   ID_CNIO                235 non-null    object\n",
      " 1   ORIGINAL ID_AP         226 non-null    object\n",
      " 2   ORIGINAL ID_NHC_BBANK  204 non-null    object\n",
      " 3   SUBTIPO HISTOLÓGICO    235 non-null    object\n",
      " 4   SERIE                  235 non-null    object\n",
      "dtypes: object(5)\n",
      "memory usage: 9.3+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "####### RECOVER MISSING DATA IF IT IS POSSIBLE #################\n",
    "# Some samples lack some data that we can recover from the identifiers file\n",
    "identificadores_df = pd.read_excel(\"Identificadores OVE_LP_MDA_RVB_serie completa.xlsx\", sheet_name='identificadores serie completa ')\n",
    "print(identificadores_df.info())\n",
    "identificadores_df.columns=identificadores_df.columns.str.strip()\n",
    "print(identificadores_df.info())\n",
    "identificadores_df.rename(columns={'ID CNIO': 'ID_CNIO'}, inplace=True)\n",
    "print(identificadores_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "603f00e1-ddfb-490e-b525-bf6ef78d00e7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       ID_CNIO ID_ORIGINAL ID_ORIGINAL_NHC_BBANK HISTOLOGY HISTOLOGY_DETAIL  \\\n",
      "22        LP23         NaN                   NaN         0              NaN   \n",
      "309       LP52         NaN                   NaN       NaN              NaN   \n",
      "310       LP54         NaN                   NaN       NaN              NaN   \n",
      "311      LP124         NaN                   NaN       NaN              NaN   \n",
      "312   03B07829         NaN                   NaN       NaN              NaN   \n",
      "313   07B23784         NaN                   NaN       NaN              NaN   \n",
      "314   10B36563         NaN                   NaN       NaN              NaN   \n",
      "315   11B23809         NaN                   NaN       NaN              NaN   \n",
      "316    11B5978         NaN                   NaN       NaN              NaN   \n",
      "317      07T25         NaN                   NaN       NaN              NaN   \n",
      "318      08T96         NaN                   NaN       NaN              NaN   \n",
      "319     09T134         NaN                   NaN       NaN              NaN   \n",
      "320      07T29         NaN                   NaN       NaN              NaN   \n",
      "321     08T107         NaN                   NaN       NaN              NaN   \n",
      "322    09T-114         NaN                   NaN       NaN              NaN   \n",
      "323  09T-117-T         NaN                   NaN       NaN              NaN   \n",
      "324    09T-119         NaN                   NaN       NaN              NaN   \n",
      "325      09T45         NaN                   NaN       NaN              NaN   \n",
      "326      08T94         NaN                   NaN       NaN              NaN   \n",
      "327     09T237         NaN                   NaN       NaN              NaN   \n",
      "328      11T82         NaN                   NaN       NaN              NaN   \n",
      "329     16T131         NaN                   NaN       NaN              NaN   \n",
      "\n",
      "     AGE GRADE FIGO FIGOL FIGOa  ... Non-MMRd_mutation TILs_score_info  \\\n",
      "22    45     1    3     0     1  ...               NaN             NaN   \n",
      "309  NaN   NaN  NaN   NaN   NaN  ...                NO             YES   \n",
      "310  NaN   NaN  NaN   NaN   NaN  ...                NO             YES   \n",
      "311  NaN   NaN  NaN   NaN   NaN  ...                NO             YES   \n",
      "312  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "313  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "314  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "315  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "316  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "317  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "318  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "319  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "320  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "321  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "322  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "323  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "324  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "325  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "326  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "327  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "328  NaN   NaN  NaN   NaN   NaN  ...               NaN             YES   \n",
      "329  NaN   NaN  NaN   NaN   NaN  ...              MSH3              NO   \n",
      "\n",
      "    TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP TILs_ep_OTTA TILs_tu_CNIO  \\\n",
      "22             NaN          NaN        NaN          NaN          NaN   \n",
      "309   YES_complete          1.0        NaN          NaN          2.0   \n",
      "310    YES_only_ep          0.0        NaN          NaN          9.0   \n",
      "311   YES_complete          2.0        NaN          NaN          3.0   \n",
      "312   YES_complete          0.0        NaN          NaN          1.0   \n",
      "313   YES_complete          0.0        NaN          NaN          2.0   \n",
      "314   YES_complete          1.0        NaN          NaN          2.0   \n",
      "315   YES_complete          2.0        NaN          NaN          2.0   \n",
      "316   YES_complete          1.0        NaN          NaN          3.0   \n",
      "317   YES_complete          0.0        NaN          NaN          0.0   \n",
      "318   YES_complete          2.0        NaN          NaN          2.0   \n",
      "319             NO          1.0        NaN          NaN          2.0   \n",
      "320             NO          9.0        NaN          NaN          9.0   \n",
      "321   YES_complete          3.0        NaN          NaN          3.0   \n",
      "322   YES_complete          3.0        NaN          NaN          3.0   \n",
      "323             NO          9.0        NaN          NaN          9.0   \n",
      "324             NO          9.0        NaN          NaN          9.0   \n",
      "325             NO          2.0        NaN          NaN          2.0   \n",
      "326   YES_complete          2.0        NaN          NaN          2.0   \n",
      "327   YES_complete          2.0        NaN          NaN          3.0   \n",
      "328             NO          0.0        NaN          NaN          2.0   \n",
      "329             NO          NaN        NaN          NaN          NaN   \n",
      "\n",
      "    TILs_tu_LP TILs_raw_ep TILs_raw_tu  \n",
      "22         NaN         NaN         NaN  \n",
      "309        NaN         1.0         9.0  \n",
      "310        NaN         0.0         NaN  \n",
      "311        NaN         6.0       185.0  \n",
      "312        NaN         0.0         2.0  \n",
      "313        NaN         0.0         9.0  \n",
      "314        NaN         1.0         9.0  \n",
      "315        NaN         4.0         5.0  \n",
      "316        NaN         1.0        47.0  \n",
      "317        NaN         0.0         0.0  \n",
      "318        NaN         6.0        14.0  \n",
      "319        NaN         NaN         NaN  \n",
      "320        NaN         NaN         NaN  \n",
      "321        NaN       105.0       146.0  \n",
      "322        NaN        22.0        52.0  \n",
      "323        NaN         NaN         NaN  \n",
      "324        NaN         NaN         NaN  \n",
      "325        NaN         NaN         NaN  \n",
      "326        NaN         4.0        19.0  \n",
      "327        NaN         5.0        27.0  \n",
      "328        NaN         NaN         NaN  \n",
      "329        NaN         NaN         NaN  \n",
      "\n",
      "[22 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "# Filter data to keep those in 'final_df' in which 'ID_ORIGINAL' and 'ID_ORIGINAL_NHC_BBANK' are NaN\n",
    "df_missing = final_df[(final_df['ID_ORIGINAL'].isna()) & (final_df['ID_ORIGINAL_NHC_BBANK'].isna())]\n",
    "print(df_missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a8857f1c-2683-4ca5-b49b-f310f8666a14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      ID_CNIO                    ORIGINAL ID_AP ORIGINAL ID_NHC_BBANK\n",
      "0        LP23                               NaN                   NaN\n",
      "1        LP52                        06B0015868               2142532\n",
      "2        LP54                        07B0030433               2258343\n",
      "3       LP124                        02B0013829                   NaN\n",
      "4    03B07829                               NaN                   NaN\n",
      "5    07B23784                               NaN                   NaN\n",
      "6    10B36563                               NaN                   NaN\n",
      "7    11B23809                               NaN                   NaN\n",
      "8     11B5978                               NaN                   NaN\n",
      "9       07T25                        8175-04/A4          BTCNIO06/111\n",
      "10      08T96        04B0008135 (F.H. ALCORCÓN)                   NaN\n",
      "11     09T134     B781787  (H,Gregorio Marañon)               B781787\n",
      "12      07T29                               NaN                   NaN\n",
      "13     08T107                               NaN                   NaN\n",
      "14    09T-114                               NaN                   NaN\n",
      "15  09T-117-T                               NaN                   NaN\n",
      "16    09T-119                               NaN                   NaN\n",
      "17      09T45                               NaN                   NaN\n",
      "18      08T94         99B5119-8 (F.H. ALCORCÓN)                   NaN\n",
      "19     09T237                        08B1440-A4                   NaN\n",
      "20      11T82        074040B5(B5;B6) (ALBACETE)                   NaN\n",
      "21     16T131  13B9434-2-7 H. U. Sant Joan Reus             NHC:51092\n"
     ]
    }
   ],
   "source": [
    "# Merge with identifiers dataFrame to obtain only the missing values\n",
    "merged_df = df_missing[['ID_CNIO']].merge(identificadores_df[['ID_CNIO', 'ORIGINAL ID_AP', 'ORIGINAL ID_NHC_BBANK']],\n",
    "                                          on='ID_CNIO', how='left')\n",
    "# Verify\n",
    "print(merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7c739965-5be2-4932-95c9-7bd9d5aa579f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Índexes of df_missing: Index([ 22, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321,\n",
      "       322, 323, 324, 325, 326, 327, 328, 329],\n",
      "      dtype='int64')\n",
      "Índexes of merged_df: RangeIndex(start=0, stop=22, step=1)\n",
      "  ID_CNIO ID_ORIGINAL ID_ORIGINAL_NHC_BBANK HISTOLOGY HISTOLOGY_DETAIL AGE  \\\n",
      "0     LP1    02B28662                   NaN         0              NaN  50   \n",
      "1     LP2    08B18876                   NaN         0              NaN  42   \n",
      "2     LP3     08B260A                   NaN         0              NaN  68   \n",
      "3     LP4    11B18858                   NaN         0              NaN  50   \n",
      "4     LP5     99B6752                   NaN         0              NaN  82   \n",
      "\n",
      "  GRADE FIGO FIGOL FIGOa  ... Non-MMRd_mutation TILs_score_info  \\\n",
      "0     2    3     0     1  ...                NO             YES   \n",
      "1     2    3     0     1  ...                NO             YES   \n",
      "2     2    5     0     2  ...                NO             YES   \n",
      "3     1    6     0     2  ...                NO             YES   \n",
      "4     1    1     0     1  ...                NO             YES   \n",
      "\n",
      "  TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP TILs_ep_OTTA TILs_tu_CNIO TILs_tu_LP  \\\n",
      "0   YES_complete          1.0        9.0          9.0          2.0        2.0   \n",
      "1   YES_complete          2.0        0.0          0.0          2.0        0.0   \n",
      "2   YES_complete          3.0        3.0          1.0          3.0        3.0   \n",
      "3   YES_complete          2.0        3.0          2.0          2.0        3.0   \n",
      "4             NO          NaN        1.0          0.0          NaN        1.0   \n",
      "\n",
      "  TILs_raw_ep TILs_raw_tu  \n",
      "0         2.0         6.0  \n",
      "1         4.0         7.0  \n",
      "2        56.0        78.0  \n",
      "3         5.0        15.0  \n",
      "4         NaN         NaN  \n",
      "\n",
      "[5 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "# Make sure the indexes are aligneated\n",
    "print(\"Índexes of df_missing:\", df_missing.index)\n",
    "print(\"Índexes of merged_df:\", merged_df.index)\n",
    "# Update values in final_df\n",
    "final_df.loc[df_missing.index, 'ID_ORIGINAL'] = merged_df['ORIGINAL ID_AP'].values\n",
    "final_df.loc[df_missing.index, 'ID_ORIGINAL_NHC_BBANK'] = merged_df['ORIGINAL ID_NHC_BBANK'].values\n",
    "# Rename the columns as required\n",
    "final_df.rename(columns={'ORIGINAL ID_AP': 'ID_ORIGINAL', 'ORIGINAL ID_NHC_BBANK': 'ID_ORIGINAL_NHC_BBANK'}, inplace=True)\n",
    "# Check the final dataFrame \n",
    "print(final_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "62d7c97d-a6ed-4e6d-8083-c1ac2774b0df",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       ID_CNIO HISTOLOGY\n",
      "300       RVB1       NaN\n",
      "307       RVB8       NaN\n",
      "308       RVB9       NaN\n",
      "309       LP52       NaN\n",
      "310       LP54       NaN\n",
      "311      LP124       NaN\n",
      "312   03B07829       NaN\n",
      "313   07B23784       NaN\n",
      "314   10B36563       NaN\n",
      "315   11B23809       NaN\n",
      "316    11B5978       NaN\n",
      "317      07T25       NaN\n",
      "318      08T96       NaN\n",
      "319     09T134       NaN\n",
      "320      07T29       NaN\n",
      "321     08T107       NaN\n",
      "322    09T-114       NaN\n",
      "323  09T-117-T       NaN\n",
      "324    09T-119       NaN\n",
      "325      09T45       NaN\n",
      "326      08T94       NaN\n",
      "327     09T237       NaN\n",
      "328      11T82       NaN\n",
      "329     16T131       NaN\n"
     ]
    }
   ],
   "source": [
    "# Remember that df dataframe was read previously. Now we need it to add the incomplete data in the HISTOLOGY column\n",
    "# for some samples. The idea is to check with are the rows with empty data and then find the value of HISTOLOGY in HIST column\n",
    "# in Noelia complete database.ods (dataframe df). If in that column the value is 'E' we must assign 0, if 'CC' we must assign 1\n",
    "# If the value contains a string with 'M' or 'mixto' we must assign 2. If it is a string without 'M' or 'mixto' must assign 3 \n",
    "# and if it is empty keep it the same.\n",
    "# print(df.info())\n",
    "df.columns=df.columns.str.strip()\n",
    "#  Filter data of 'final_df' in which 'HISTOLOGY' value is NaN\n",
    "final_missing = final_df[(final_df['HISTOLOGY'].isna())]\n",
    "print(final_missing[['ID_CNIO','HISTOLOGY']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "16a1e16b-8b7f-4cc1-a702-84463b72f35e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      ID_CNIO         HIST\n",
      "0        RVB1  NA (Colon?)\n",
      "1        RVB8          NaN\n",
      "2        RVB9          NaN\n",
      "3        LP52            E\n",
      "4        LP54            E\n",
      "5       LP124           CC\n",
      "6    03B07829          NaN\n",
      "7    07B23784          NaN\n",
      "8    10B36563          NaN\n",
      "9    11B23809          NaN\n",
      "10    11B5978          NaN\n",
      "11      07T25           CC\n",
      "12      08T96           CC\n",
      "13     09T134           CC\n",
      "14      07T29          NaN\n",
      "15     08T107          NaN\n",
      "16    09T-114          NaN\n",
      "17  09T-117-T          NaN\n",
      "18    09T-119          NaN\n",
      "19      09T45          NaN\n",
      "20      08T94            E\n",
      "21     09T237            E\n",
      "22      11T82            E\n",
      "23     16T131            E\n",
      "Índexes of df_missing: Index([300, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319,\n",
      "       320, 321, 322, 323, 324, 325, 326, 327, 328, 329],\n",
      "      dtype='int64')\n",
      "Índexes of merged_df: RangeIndex(start=0, stop=24, step=1)\n"
     ]
    }
   ],
   "source": [
    "# Merge to obtain missing values from the df dataframe (it contains genomic+TILs data)\n",
    "merged_df2 = final_missing[['ID_CNIO']].merge(df[['ID_CNIO', 'HIST']],\n",
    "                                          on='ID_CNIO', how='left')\n",
    "# Verify\n",
    "print(merged_df2)\n",
    "print(\"Índexes of df_missing:\", final_missing.index)\n",
    "print(\"Índexes of merged_df:\", merged_df2.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2faf9007-872c-43e7-b91f-fe08d18c99cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  ID_CNIO ID_ORIGINAL ID_ORIGINAL_NHC_BBANK HISTOLOGY HISTOLOGY_DETAIL AGE  \\\n",
      "0     LP1    02B28662                   NaN         0              NaN  50   \n",
      "1     LP2    08B18876                   NaN         0              NaN  42   \n",
      "2     LP3     08B260A                   NaN         0              NaN  68   \n",
      "3     LP4    11B18858                   NaN         0              NaN  50   \n",
      "4     LP5     99B6752                   NaN         0              NaN  82   \n",
      "\n",
      "  GRADE FIGO FIGOL FIGOa  ... Non-MMRd_mutation TILs_score_info  \\\n",
      "0     2    3     0     1  ...                NO             YES   \n",
      "1     2    3     0     1  ...                NO             YES   \n",
      "2     2    5     0     2  ...                NO             YES   \n",
      "3     1    6     0     2  ...                NO             YES   \n",
      "4     1    1     0     1  ...                NO             YES   \n",
      "\n",
      "  TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP TILs_ep_OTTA TILs_tu_CNIO TILs_tu_LP  \\\n",
      "0   YES_complete          1.0        9.0          9.0          2.0        2.0   \n",
      "1   YES_complete          2.0        0.0          0.0          2.0        0.0   \n",
      "2   YES_complete          3.0        3.0          1.0          3.0        3.0   \n",
      "3   YES_complete          2.0        3.0          2.0          2.0        3.0   \n",
      "4             NO          NaN        1.0          0.0          NaN        1.0   \n",
      "\n",
      "  TILs_raw_ep TILs_raw_tu  \n",
      "0         2.0         6.0  \n",
      "1         4.0         7.0  \n",
      "2        56.0        78.0  \n",
      "3         5.0        15.0  \n",
      "4         NaN         NaN  \n",
      "\n",
      "[5 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "# Update values in final_df\n",
    "final_df.loc[final_missing.index, 'HISTOLOGY'] = merged_df2['HIST'].values\n",
    "print(final_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d0eada64-7a29-413f-a79c-72f642219fe6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  ID_CNIO ID_ORIGINAL ID_ORIGINAL_NHC_BBANK HISTOLOGY HISTOLOGY_DETAIL AGE  \\\n",
      "0     LP1    02B28662                   NaN         0              NaN  50   \n",
      "1     LP2    08B18876                   NaN         0              NaN  42   \n",
      "2     LP3     08B260A                   NaN         0              NaN  68   \n",
      "3     LP4    11B18858                   NaN         0              NaN  50   \n",
      "4     LP5     99B6752                   NaN         0              NaN  82   \n",
      "\n",
      "  GRADE FIGO FIGOL FIGOa  ... Non-MMRd_mutation TILs_score_info  \\\n",
      "0     2    3     0     1  ...                NO             YES   \n",
      "1     2    3     0     1  ...                NO             YES   \n",
      "2     2    5     0     2  ...                NO             YES   \n",
      "3     1    6     0     2  ...                NO             YES   \n",
      "4     1    1     0     1  ...                NO             YES   \n",
      "\n",
      "  TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP TILs_ep_OTTA TILs_tu_CNIO TILs_tu_LP  \\\n",
      "0   YES_complete          1.0        9.0          9.0          2.0        2.0   \n",
      "1   YES_complete          2.0        0.0          0.0          2.0        0.0   \n",
      "2   YES_complete          3.0        3.0          1.0          3.0        3.0   \n",
      "3   YES_complete          2.0        3.0          2.0          2.0        3.0   \n",
      "4             NO          NaN        1.0          0.0          NaN        1.0   \n",
      "\n",
      "  TILs_raw_ep TILs_raw_tu  \n",
      "0         2.0         6.0  \n",
      "1         4.0         7.0  \n",
      "2        56.0        78.0  \n",
      "3         5.0        15.0  \n",
      "4         NaN         NaN  \n",
      "\n",
      "[5 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "#  Rename the columns as required\n",
    "final_df.rename(columns={'HIST': 'HISTOLOGY'}, inplace=True)\n",
    "# Check\n",
    "print(final_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "29f2da13-43ec-4d34-ba69-149739ef6cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "object\n"
     ]
    }
   ],
   "source": [
    "# I analyse the type of data we have in the 'HISTOLOGY' column.\n",
    "print(final_df['HISTOLOGY'].dtype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8464256c-4ce9-4bd8-9c73-8002470bdc13",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    ID_CNIO                       ID_ORIGINAL ID_ORIGINAL_NHC_BBANK  \\\n",
      "0       LP1                          02B28662                   NaN   \n",
      "1       LP2                          08B18876                   NaN   \n",
      "2       LP3                           08B260A                   NaN   \n",
      "3       LP4                          11B18858                   NaN   \n",
      "4       LP5                           99B6752                   NaN   \n",
      "..      ...                               ...                   ...   \n",
      "325   09T45                               NaN                   NaN   \n",
      "326   08T94         99B5119-8 (F.H. ALCORCÓN)                   NaN   \n",
      "327  09T237                        08B1440-A4                   NaN   \n",
      "328   11T82        074040B5(B5;B6) (ALBACETE)                   NaN   \n",
      "329  16T131  13B9434-2-7 H. U. Sant Joan Reus             NHC:51092   \n",
      "\n",
      "     HISTOLOGY HISTOLOGY_DETAIL  AGE GRADE FIGO FIGOL FIGOa  ...  \\\n",
      "0          0.0              NaN   50     2    3     0     1  ...   \n",
      "1          0.0              NaN   42     2    3     0     1  ...   \n",
      "2          0.0              NaN   68     2    5     0     2  ...   \n",
      "3          0.0              NaN   50     1    6     0     2  ...   \n",
      "4          0.0              NaN   82     1    1     0     1  ...   \n",
      "..         ...              ...  ...   ...  ...   ...   ...  ...   \n",
      "325        NaN              NaN  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "326        0.0              NaN  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "327        0.0              NaN  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "328        0.0              NaN  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "329        0.0              NaN  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "\n",
      "    Non-MMRd_mutation TILs_score_info TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP  \\\n",
      "0                  NO             YES   YES_complete          1.0        9.0   \n",
      "1                  NO             YES   YES_complete          2.0        0.0   \n",
      "2                  NO             YES   YES_complete          3.0        3.0   \n",
      "3                  NO             YES   YES_complete          2.0        3.0   \n",
      "4                  NO             YES             NO          NaN        1.0   \n",
      "..                ...             ...            ...          ...        ...   \n",
      "325               NaN             YES             NO          2.0        NaN   \n",
      "326               NaN             YES   YES_complete          2.0        NaN   \n",
      "327               NaN             YES   YES_complete          2.0        NaN   \n",
      "328               NaN             YES             NO          0.0        NaN   \n",
      "329              MSH3              NO             NO          NaN        NaN   \n",
      "\n",
      "    TILs_ep_OTTA TILs_tu_CNIO TILs_tu_LP TILs_raw_ep TILs_raw_tu  \n",
      "0            9.0          2.0        2.0         2.0         6.0  \n",
      "1            0.0          2.0        0.0         4.0         7.0  \n",
      "2            1.0          3.0        3.0        56.0        78.0  \n",
      "3            2.0          2.0        3.0         5.0        15.0  \n",
      "4            0.0          NaN        1.0         NaN         NaN  \n",
      "..           ...          ...        ...         ...         ...  \n",
      "325          NaN          2.0        NaN         NaN         NaN  \n",
      "326          NaN          2.0        NaN         4.0        19.0  \n",
      "327          NaN          3.0        NaN         5.0        27.0  \n",
      "328          NaN          2.0        NaN         NaN         NaN  \n",
      "329          NaN          NaN        NaN         NaN         NaN  \n",
      "\n",
      "[330 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "# Now I have to change the values with the new coding without modifiying the rows that already have the right codification.\n",
    "# Function to transform the values in the 'HISTOLOGY'column according to the definitive coding.\n",
    "\n",
    "def transform_histology(value):\n",
    "    if pd.isna(value) or value in [0, 1, 2, 3]:\n",
    "        return value\n",
    "    if isinstance(value, str):  # Verifica si es una cadena\n",
    "        value = value.strip()  # Elimina espacios en blanco\n",
    "        if value == 'E':\n",
    "            return 0\n",
    "        elif value == 'CC':\n",
    "            return 1\n",
    "        elif 'M' in value or 'mixto' in value.lower() or 'mixed' in value.lower():\n",
    "            return 2\n",
    "        else:\n",
    "            return 3\n",
    "    return value  # Si es un valor numérico, lo devolvemos tal cual\n",
    "    \n",
    "# Apply the function to the 'HISTOLOGY' column\n",
    "final_df['HISTOLOGY'] = final_df['HISTOLOGY'].apply(transform_histology)\n",
    "\n",
    "# Show the final dataframe\n",
    "print(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e11c4618-e69f-4fe0-a98f-a32612c451a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    ID_CNIO                       ID_ORIGINAL ID_ORIGINAL_NHC_BBANK  \\\n",
      "0       LP1                          02B28662                   NaN   \n",
      "1       LP2                          08B18876                   NaN   \n",
      "2       LP3                           08B260A                   NaN   \n",
      "3       LP4                          11B18858                   NaN   \n",
      "4       LP5                           99B6752                   NaN   \n",
      "..      ...                               ...                   ...   \n",
      "325   09T45                               NaN                   NaN   \n",
      "326   08T94         99B5119-8 (F.H. ALCORCÓN)                   NaN   \n",
      "327  09T237                        08B1440-A4                   NaN   \n",
      "328   11T82        074040B5(B5;B6) (ALBACETE)                   NaN   \n",
      "329  16T131  13B9434-2-7 H. U. Sant Joan Reus             NHC:51092   \n",
      "\n",
      "     HISTOLOGY HISTOLOGY_DETAIL  AGE GRADE FIGO FIGOL FIGOa  ...  \\\n",
      "0          0.0         No mixed   50     2    3     0     1  ...   \n",
      "1          0.0         No mixed   42     2    3     0     1  ...   \n",
      "2          0.0         No mixed   68     2    5     0     2  ...   \n",
      "3          0.0         No mixed   50     1    6     0     2  ...   \n",
      "4          0.0         No mixed   82     1    1     0     1  ...   \n",
      "..         ...              ...  ...   ...  ...   ...   ...  ...   \n",
      "325        NaN              NaN  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "326        0.0         No mixed  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "327        0.0         No mixed  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "328        0.0         No mixed  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "329        0.0         No mixed  NaN   NaN  NaN   NaN   NaN  ...   \n",
      "\n",
      "    Non-MMRd_mutation TILs_score_info TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP  \\\n",
      "0                  NO             YES   YES_complete          1.0        9.0   \n",
      "1                  NO             YES   YES_complete          2.0        0.0   \n",
      "2                  NO             YES   YES_complete          3.0        3.0   \n",
      "3                  NO             YES   YES_complete          2.0        3.0   \n",
      "4                  NO             YES             NO          NaN        1.0   \n",
      "..                ...             ...            ...          ...        ...   \n",
      "325               NaN             YES             NO          2.0        NaN   \n",
      "326               NaN             YES   YES_complete          2.0        NaN   \n",
      "327               NaN             YES   YES_complete          2.0        NaN   \n",
      "328               NaN             YES             NO          0.0        NaN   \n",
      "329              MSH3              NO             NO          NaN        NaN   \n",
      "\n",
      "    TILs_ep_OTTA TILs_tu_CNIO TILs_tu_LP TILs_raw_ep TILs_raw_tu  \n",
      "0            9.0          2.0        2.0         2.0         6.0  \n",
      "1            0.0          2.0        0.0         4.0         7.0  \n",
      "2            1.0          3.0        3.0        56.0        78.0  \n",
      "3            2.0          2.0        3.0         5.0        15.0  \n",
      "4            0.0          NaN        1.0         NaN         NaN  \n",
      "..           ...          ...        ...         ...         ...  \n",
      "325          NaN          2.0        NaN         NaN         NaN  \n",
      "326          NaN          2.0        NaN         4.0        19.0  \n",
      "327          NaN          3.0        NaN         5.0        27.0  \n",
      "328          NaN          2.0        NaN         NaN         NaN  \n",
      "329          NaN          NaN        NaN         NaN         NaN  \n",
      "\n",
      "[330 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "# Update 'HISTOLOGY_DETAIL' column\n",
    "# Add the value 'Mixed E+CC' in 'HISTOLOGY_DETAIL' for the row with ID_CNIO='RVB3'\n",
    "final_df.loc[final_df['ID_CNIO'] == 'RVB3', 'HISTOLOGY_DETAIL'] = 'Mixed E+CC'\n",
    "\n",
    "# Update 'HISTOLOGY_DETAIL' values according to the 'HISTOLOGY' values\n",
    "def update_histology_detail(row):\n",
    "    if pd.isna(row['HISTOLOGY']):  # If 'HISTOLOGY' is NaN, we keep it'\n",
    "        return row['HISTOLOGY_DETAIL']\n",
    "    elif row['HISTOLOGY'] in [0, 1, 3]:  # If it is 0, 1 o 3, assign 'No mixed'\n",
    "        return 'No mixed'\n",
    "    else:  # For the rest of the cases we keep the original data\n",
    "        return row['HISTOLOGY_DETAIL']\n",
    "\n",
    "# Apply the function to the column\n",
    "final_df['HISTOLOGY_DETAIL'] = final_df.apply(update_histology_detail, axis=1)\n",
    "\n",
    "# Show the final dataframe\n",
    "print(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5b3b8c1f-16fe-453e-8f45-2dfd208b01b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0        2\n",
      "1        2\n",
      "2        2\n",
      "3        1\n",
      "4        1\n",
      "      ... \n",
      "325    NaN\n",
      "326    NaN\n",
      "327    NaN\n",
      "328    NaN\n",
      "329    NaN\n",
      "Name: GRADE, Length: 330, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# Some more modifications to keep the final coding\n",
    "# GRADE column: I change the value of RVB4 'ovario:3; ENDOMETRIO: 1'for 3;1 in order to mantain the coding\n",
    "# Additionally I transform all values to integer type or a list of integers (if they are split by a ;), because there are different\n",
    "# among cohorts(float,string).\n",
    "final_df.loc[final_df['ID_CNIO']=='RVB4','GRADE']='3;1'\n",
    "# Function to convert the data. Previously I tranform all data to string type to avoid errors when using the function.\n",
    "#final_df['GRADE'] = final_df['GRADE'].astype(str)\n",
    "def convert_value(value):\n",
    "    \"\"\"Convert a string split by ';' into a list of integers or a single integer.\"\"\"\n",
    "    if pd.notna(value) and value != '':\n",
    "        # Convertir el valor a cadena de texto si no lo es\n",
    "        value = str(value)\n",
    "        if ';' in value:\n",
    "            # Convertir los números separados por ';' en una lista de enteros\n",
    "            try:\n",
    "                return [int(part) for part in value.split(';')]\n",
    "            except ValueError:\n",
    "                return np.nan\n",
    "        else:\n",
    "            try:\n",
    "                # Convertir un solo valor en un entero\n",
    "                return int(value)\n",
    "            except ValueError:\n",
    "                return np.nan\n",
    "    return np.nan  # Mantener valores NaN\n",
    "    \n",
    "# Apply the function to the column\n",
    "final_df['GRADE'] = final_df['GRADE'].apply(convert_value)\n",
    "print(final_df['GRADE'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a6ccfb2f-7566-4d14-8115-e27897bd422b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill empty cells with NO in Clinical_info, Genomic_info and TILs_score_info\n",
    "final_df['Clinical_info']=final_df['Clinical_info'].fillna('NO')\n",
    "final_df['Genomic_info']=final_df['Genomic_info'].fillna('NO')\n",
    "final_df['TILs_score_info']=final_df['TILs_score_info'].fillna('NO')\n",
    "# final_df.to_excel(\"Samples_all_info.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4beb0cd1-76ad-4664-9571-9c6a0da55a83",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 330 entries, 0 to 329\n",
      "Data columns (total 56 columns):\n",
      " #   Column                     Non-Null Count  Dtype  \n",
      "---  ------                     --------------  -----  \n",
      " 0   ID_CNIO                    330 non-null    object \n",
      " 1   ID_ORIGINAL                309 non-null    object \n",
      " 2   ID_ORIGINAL_NHC_BBANK      288 non-null    object \n",
      " 3   HISTOLOGY                  317 non-null    float64\n",
      " 4   HISTOLOGY_DETAIL           317 non-null    object \n",
      " 5   AGE                        309 non-null    object \n",
      " 6   GRADE                      243 non-null    object \n",
      " 7   FIGO                       275 non-null    object \n",
      " 8   FIGOL                      276 non-null    object \n",
      " 9   FIGOa                      275 non-null    object \n",
      " 10  NEOADJUVANT_TREATMENT      233 non-null    object \n",
      " 11  ADJUVANT_TREATMENT         279 non-null    object \n",
      " 12  TYPE_ADJUVANT              197 non-null    object \n",
      " 13  OTHER_ADJ_TREAT            14 non-null     object \n",
      " 14  RESIDUAL                   224 non-null    object \n",
      " 15  RESIDUALa                  224 non-null    object \n",
      " 16  OS_CNIO                    284 non-null    object \n",
      " 17  VITAL STATUS               289 non-null    object \n",
      " 18  MMR STATUS                 179 non-null    object \n",
      " 19  BRCA STATUS                169 non-null    object \n",
      " 20  FAMILIAL                   266 non-null    object \n",
      " 21  DATE_AGE                   129 non-null    object \n",
      " 22  DATE_OS                    308 non-null    object \n",
      " 23  PARTIAL DATE DIAGNOSIS     61 non-null     object \n",
      " 24  PARTIAL DATE DEATH_LASTv   61 non-null     object \n",
      " 25  Clinical_info              330 non-null    object \n",
      " 26  IHC_defective_MMR_protein  232 non-null    object \n",
      " 27  IHC pattern                232 non-null    object \n",
      " 28  IHC_MMR_class              232 non-null    object \n",
      " 29  PCR                        198 non-null    object \n",
      " 30  Evidence score for MMRd    31 non-null     float64\n",
      " 31  IHCd                       171 non-null    object \n",
      " 32  PCRd                       172 non-null    object \n",
      " 33  QC_filters                 205 non-null    object \n",
      " 34  Genomic_info               330 non-null    object \n",
      " 35  MMR_final_status           205 non-null    object \n",
      " 36  CLASSd                     173 non-null    object \n",
      " 37  MSI_sensor2                173 non-null    float64\n",
      " 38  TMB                        173 non-null    float64\n",
      " 39  ID2                        170 non-null    float64\n",
      " 40  ID7                        170 non-null    float64\n",
      " 41  SUM ID2+ID7                173 non-null    float64\n",
      " 42  CNV                        173 non-null    float64\n",
      " 43  %genome_altered            173 non-null    float64\n",
      " 44  Pol_mutation               173 non-null    object \n",
      " 45  MMRd_mutation              173 non-null    object \n",
      " 46  Non-MMRd_mutation          173 non-null    object \n",
      " 47  TILs_score_info            330 non-null    object \n",
      " 48  TILs_CNIO_info             235 non-null    object \n",
      " 49  TILs_ep_CNIO               211 non-null    float64\n",
      " 50  TILs_ep_LP                 127 non-null    float64\n",
      " 51  TILs_ep_OTTA               109 non-null    float64\n",
      " 52  TILs_tu_CNIO               210 non-null    float64\n",
      " 53  TILs_tu_LP                 88 non-null     float64\n",
      " 54  TILs_raw_ep                196 non-null    float64\n",
      " 55  TILs_raw_tu                197 non-null    float64\n",
      "dtypes: float64(16), object(40)\n",
      "memory usage: 144.5+ KB\n",
      "None\n",
      "  ID_CNIO ID_ORIGINAL ID_ORIGINAL_NHC_BBANK  HISTOLOGY HISTOLOGY_DETAIL AGE  \\\n",
      "0     LP1    02B28662                   NaN        0.0         No mixed  50   \n",
      "1     LP2    08B18876                   NaN        0.0         No mixed  42   \n",
      "2     LP3     08B260A                   NaN        0.0         No mixed  68   \n",
      "3     LP4    11B18858                   NaN        0.0         No mixed  50   \n",
      "4     LP5     99B6752                   NaN        0.0         No mixed  82   \n",
      "\n",
      "  GRADE FIGO FIGOL FIGOa  ... Non-MMRd_mutation TILs_score_info  \\\n",
      "0     2    3     0     1  ...                NO             YES   \n",
      "1     2    3     0     1  ...                NO             YES   \n",
      "2     2    5     0     2  ...                NO             YES   \n",
      "3     1    6     0     2  ...                NO             YES   \n",
      "4     1    1     0     1  ...                NO             YES   \n",
      "\n",
      "  TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP TILs_ep_OTTA TILs_tu_CNIO TILs_tu_LP  \\\n",
      "0   YES_complete          1.0        9.0          9.0          2.0        2.0   \n",
      "1   YES_complete          2.0        0.0          0.0          2.0        0.0   \n",
      "2   YES_complete          3.0        3.0          1.0          3.0        3.0   \n",
      "3   YES_complete          2.0        3.0          2.0          2.0        3.0   \n",
      "4             NO          NaN        1.0          0.0          NaN        1.0   \n",
      "\n",
      "  TILs_raw_ep TILs_raw_tu  \n",
      "0         2.0         6.0  \n",
      "1         4.0         7.0  \n",
      "2        56.0        78.0  \n",
      "3         5.0        15.0  \n",
      "4         NaN         NaN  \n",
      "\n",
      "[5 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "print(final_df.info())\n",
    "print(final_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ab51f35f-1418-4e75-8932-68fe8b963896",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['0' '1' 'nan']\n",
      "['0' '1' 'NA']\n"
     ]
    }
   ],
   "source": [
    "# Convert all values in the 'VITAL_STATUS' column to strings, remove extra spaces, and replace NaN values with 'NA'\n",
    "final_df['VITAL STATUS'] = final_df['VITAL STATUS'].astype(str).str.strip()\n",
    "\n",
    "# Verify the unique values to ensure there are no extra spaces\n",
    "print(final_df['VITAL STATUS'].unique())\n",
    "\n",
    "# Replace 'nan' (NaN values converted to string) with 'NA'\n",
    "final_df['VITAL STATUS'] = final_df['VITAL STATUS'].replace('nan', 'NA')\n",
    "print(final_df['VITAL STATUS'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b4fa05b9-13bc-4298-8e21-29e26befbc46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in which 'Genomic_info', 'Clinical_info', and 'TILs_score_info' is 'YES': 167\n",
      "Number of rows in which 'Genomic_info', 'Clinical_info', and 'TILs_CNIO_info' is YES (incomplete or complete for TILs): 158\n",
      "Number of rows in which 'Genomic_info', 'Clinical_info', and 'TILs_CNIO_info' is YES (complete for TILs): 157\n"
     ]
    }
   ],
   "source": [
    "# Counts to know how many samples have all data type\n",
    "# Filter rows in which 'Genomic_info' is 'YES'\n",
    "genomic_yes = final_df[final_df['Genomic_info'] == 'YES']\n",
    "\n",
    "# Fill the rows with Genomic_info that are 'YES' in 'Clinical_info' and 'TILs_score_info' too\n",
    "yes_count = genomic_yes[(genomic_yes['Clinical_info'] == 'YES') & (genomic_yes['TILs_score_info'] == 'YES')]\n",
    "yes_count_cnio=genomic_yes[(genomic_yes['Clinical_info'] == 'YES') & (genomic_yes['TILs_CNIO_info'] != 'NO')]\n",
    "yes_count_cnio_complete=genomic_yes[(genomic_yes['Clinical_info'] == 'YES') & (genomic_yes['TILs_CNIO_info'] == 'YES_complete')]\n",
    "# Count the number of rows\n",
    "count_yes_yes_yes = yes_count.shape[0]\n",
    "count_yes_cnio= yes_count_cnio.shape[0]\n",
    "count_yes_cnio_complete=yes_count_cnio_complete.shape[0]\n",
    "# Save the results in a new DataFrame and export it to an excel file\n",
    "new_df = yes_count.copy()\n",
    "# Show the results\n",
    "print(f\"Number of rows in which 'Genomic_info', 'Clinical_info', and 'TILs_score_info' is 'YES': {count_yes_yes_yes}\") #167\n",
    "print(f\"Number of rows in which 'Genomic_info', 'Clinical_info', and 'TILs_CNIO_info' is YES (incomplete or complete for TILs): {count_yes_cnio}\") \n",
    "print(f\"Number of rows in which 'Genomic_info', 'Clinical_info', and 'TILs_CNIO_info' is YES (complete for TILs): {count_yes_cnio_complete}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "923618b6-d829-42a7-96a5-005a09b1b7bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with 'YES' in 'Clinical_info' and 'TILs_info': 212\n",
      "Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info': 169\n",
      "Number of rows with 'YES' in 'Genomic_info' and 'TILs_info': 170\n"
     ]
    }
   ],
   "source": [
    "# Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' regardless of 'Genomic_info'\n",
    "clinical_tils_yes = final_df[(final_df['Clinical_info'] == 'YES') & (final_df['TILs_score_info'] == 'YES')]\n",
    "count_clinical_tils_yes = clinical_tils_yes.shape[0]\n",
    "\n",
    "# Count how many rows have 'YES' in 'Clinical_info' and 'Genomic_info' regardless of 'TILs_info'\n",
    "clinical_genomic_yes = final_df[(final_df['Clinical_info'] == 'YES') & (final_df['Genomic_info'] == 'YES')]\n",
    "count_clinical_genomic_yes = clinical_genomic_yes.shape[0]\n",
    "\n",
    "# Count how many rows have 'YES' in 'Genomic_info' and 'TILs_info' regardless of 'Clinical_info'\n",
    "genomic_tils_yes = final_df[(final_df['Genomic_info'] == 'YES') & (final_df['TILs_score_info'] == 'YES')]\n",
    "count_genomic_tils_yes = genomic_tils_yes.shape[0]\n",
    "\n",
    "# Show the results\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'TILs_info': {count_clinical_tils_yes}\")\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info': {count_clinical_genomic_yes}\")\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info' and 'TILs_info': {count_genomic_tils_yes}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "14e6738d-50e0-4415-a2bf-be151fdb4ef6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are MMRd: 24\n",
      "Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are MMRp: 169\n",
      "Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are MMRd: 21\n",
      "Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are MMRp: 148\n",
      "Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are MMRd: 20\n",
      "Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are MMRp: 150\n",
      "Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are MMRd: 20\n",
      "Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are MMRp: 147\n"
     ]
    }
   ],
   "source": [
    "# Original filtering without excluding index 0\n",
    "clinical_tils_yes = final_df[(final_df['Clinical_info'] == 'YES') & (final_df['TILs_score_info'] == 'YES')]\n",
    "clinical_genomic_yes = final_df[(final_df['Clinical_info'] == 'YES') & (final_df['Genomic_info'] == 'YES')]\n",
    "genomic_tils_yes = final_df[(final_df['Genomic_info'] == 'YES') & (final_df['TILs_score_info'] == 'YES')]\n",
    "all_yes = final_df[(final_df['Clinical_info'] == 'YES') & (final_df['Genomic_info'] == 'YES') & (final_df['TILs_score_info'] == 'YES')]\n",
    "\n",
    "# Count how many are MMRd and MMRp in each group\n",
    "\n",
    "# For Clinical_info and TILs_info\n",
    "clinical_tils_MMRd = clinical_tils_yes[clinical_tils_yes['MMR_final_status'] == 'MMRd'].shape[0]\n",
    "clinical_tils_MMRp = clinical_tils_yes[clinical_tils_yes['MMR_final_status'] == 'MMRp'].shape[0]\n",
    "\n",
    "# For Clinical_info and Genomic_info\n",
    "clinical_genomic_MMRd = clinical_genomic_yes[clinical_genomic_yes['MMR_final_status'] == 'MMRd'].shape[0]\n",
    "clinical_genomic_MMRp = clinical_genomic_yes[clinical_genomic_yes['MMR_final_status'] == 'MMRp'].shape[0]\n",
    "\n",
    "# For Genomic_info and TILs_score_info\n",
    "genomic_tils_MMRd = genomic_tils_yes[genomic_tils_yes['MMR_final_status'] == 'MMRd'].shape[0]\n",
    "genomic_tils_MMRp = genomic_tils_yes[genomic_tils_yes['MMR_final_status'] == 'MMRp'].shape[0]\n",
    "\n",
    "# For Genomic_info, Clinical_info, and TILs_score_info\n",
    "all_yes_MMRd = all_yes[all_yes['MMR_final_status'] == 'MMRd'].shape[0]\n",
    "all_yes_MMRp = all_yes[all_yes['MMR_final_status'] == 'MMRp'].shape[0]\n",
    "\n",
    "# Display results\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are MMRd: {clinical_tils_MMRd}\")\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are MMRp: {clinical_tils_MMRp}\")\n",
    "\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are MMRd: {clinical_genomic_MMRd}\")\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are MMRp: {clinical_genomic_MMRp}\")\n",
    "\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are MMRd: {genomic_tils_MMRd}\")\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are MMRp: {genomic_tils_MMRp}\")\n",
    "\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are MMRd: {all_yes_MMRd}\")\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are MMRp: {all_yes_MMRp}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "03ac0d1a-cdd1-491c-8a36-f6d7a480d3a1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are E: 103\n",
      "Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are CC: 100\n",
      "Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are E: 90\n",
      "Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are CC: 78\n",
      "Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are E: 91\n",
      "Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are CC: 78\n",
      "Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are E: 89\n",
      "Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are CC: 77\n"
     ]
    }
   ],
   "source": [
    "# Count how many are E (0) and CC(1) in each group\n",
    "\n",
    "# For Clinical_info and TILs_info\n",
    "clinical_tils_E = clinical_tils_yes[clinical_tils_yes['HISTOLOGY'] == 0].shape[0]\n",
    "clinical_tils_CC = clinical_tils_yes[clinical_tils_yes['HISTOLOGY'] == 1].shape[0]\n",
    "\n",
    "# For Clinical_info and Genomic_info\n",
    "clinical_genomic_E = clinical_genomic_yes[clinical_genomic_yes['HISTOLOGY'] == 0].shape[0]\n",
    "clinical_genomic_CC = clinical_genomic_yes[clinical_genomic_yes['HISTOLOGY'] == 1].shape[0]\n",
    "\n",
    "# For Genomic_info and TILs_info\n",
    "genomic_tils_E = genomic_tils_yes[genomic_tils_yes['HISTOLOGY'] == 0].shape[0]\n",
    "genomic_tils_CC = genomic_tils_yes[genomic_tils_yes['HISTOLOGY'] == 1].shape[0]\n",
    "\n",
    "# For Genomic_info, Clinical_info, and TILs_info\n",
    "all_yes_E = all_yes[all_yes['HISTOLOGY'] == 0].shape[0]\n",
    "all_yes_CC = all_yes[all_yes['HISTOLOGY'] == 1].shape[0]\n",
    "\n",
    "# Display results\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are E: {clinical_tils_E}\")\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'TILs_info' that are CC: {clinical_tils_CC}\")\n",
    "\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are E: {clinical_genomic_E}\")\n",
    "print(f\"Number of rows with 'YES' in 'Clinical_info' and 'Genomic_info' that are CC: {clinical_genomic_CC}\")\n",
    "\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are E: {genomic_tils_E}\")\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info' and 'TILs_info' that are CC: {genomic_tils_CC}\")\n",
    "\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are E: {all_yes_E}\")\n",
    "print(f\"Number of rows with 'YES' in 'Genomic_info', 'Clinical_info', and 'TILs_info' that are CC: {all_yes_CC}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2a496aaa-87b1-4fb8-9075-058416edd77c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 167 entries, 0 to 305\n",
      "Data columns (total 56 columns):\n",
      " #   Column                     Non-Null Count  Dtype  \n",
      "---  ------                     --------------  -----  \n",
      " 0   ID_CNIO                    167 non-null    object \n",
      " 1   ID_ORIGINAL                163 non-null    object \n",
      " 2   ID_ORIGINAL_NHC_BBANK      147 non-null    object \n",
      " 3   HISTOLOGY                  167 non-null    float64\n",
      " 4   HISTOLOGY_DETAIL           167 non-null    object \n",
      " 5   AGE                        167 non-null    object \n",
      " 6   GRADE                      132 non-null    object \n",
      " 7   FIGO                       161 non-null    object \n",
      " 8   FIGOL                      161 non-null    object \n",
      " 9   FIGOa                      161 non-null    object \n",
      " 10  NEOADJUVANT_TREATMENT      111 non-null    object \n",
      " 11  ADJUVANT_TREATMENT         152 non-null    object \n",
      " 12  TYPE_ADJUVANT              121 non-null    object \n",
      " 13  OTHER_ADJ_TREAT            7 non-null      object \n",
      " 14  RESIDUAL                   132 non-null    object \n",
      " 15  RESIDUALa                  132 non-null    object \n",
      " 16  OS_CNIO                    156 non-null    object \n",
      " 17  VITAL STATUS               167 non-null    object \n",
      " 18  MMR STATUS                 134 non-null    object \n",
      " 19  BRCA STATUS                129 non-null    object \n",
      " 20  FAMILIAL                   146 non-null    object \n",
      " 21  DATE_AGE                   32 non-null     object \n",
      " 22  DATE_OS                    167 non-null    object \n",
      " 23  PARTIAL DATE DIAGNOSIS     53 non-null     object \n",
      " 24  PARTIAL DATE DEATH_LASTv   53 non-null     object \n",
      " 25  Clinical_info              167 non-null    object \n",
      " 26  IHC_defective_MMR_protein  164 non-null    object \n",
      " 27  IHC pattern                164 non-null    object \n",
      " 28  IHC_MMR_class              164 non-null    object \n",
      " 29  PCR                        166 non-null    object \n",
      " 30  Evidence score for MMRd    22 non-null     float64\n",
      " 31  IHCd                       166 non-null    object \n",
      " 32  PCRd                       166 non-null    object \n",
      " 33  QC_filters                 167 non-null    object \n",
      " 34  Genomic_info               167 non-null    object \n",
      " 35  MMR_final_status           167 non-null    object \n",
      " 36  CLASSd                     167 non-null    object \n",
      " 37  MSI_sensor2                167 non-null    float64\n",
      " 38  TMB                        167 non-null    float64\n",
      " 39  ID2                        164 non-null    float64\n",
      " 40  ID7                        164 non-null    float64\n",
      " 41  SUM ID2+ID7                167 non-null    float64\n",
      " 42  CNV                        167 non-null    float64\n",
      " 43  %genome_altered            167 non-null    float64\n",
      " 44  Pol_mutation               167 non-null    object \n",
      " 45  MMRd_mutation              167 non-null    object \n",
      " 46  Non-MMRd_mutation          167 non-null    object \n",
      " 47  TILs_score_info            167 non-null    object \n",
      " 48  TILs_CNIO_info             167 non-null    object \n",
      " 49  TILs_ep_CNIO               164 non-null    float64\n",
      " 50  TILs_ep_LP                 96 non-null     float64\n",
      " 51  TILs_ep_OTTA               78 non-null     float64\n",
      " 52  TILs_tu_CNIO               163 non-null    float64\n",
      " 53  TILs_tu_LP                 66 non-null     float64\n",
      " 54  TILs_raw_ep                157 non-null    float64\n",
      " 55  TILs_raw_tu                158 non-null    float64\n",
      "dtypes: float64(16), object(40)\n",
      "memory usage: 74.4+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(new_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fc2563b0-c409-4949-84a9-323bca99d595",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             ID_CNIO  \\\n",
      "0                       Unique CNIO tumor identifier   \n",
      "1  LP: ID for samples from La Paz Hospital; OVE: ...   \n",
      "2                                                LP1   \n",
      "3                                                LP2   \n",
      "4                                                LP3   \n",
      "\n",
      "                                         ID_ORIGINAL  \\\n",
      "0  Unique identifier from the hospital to the pat...   \n",
      "1                                     Pathology code   \n",
      "2                                           02B28662   \n",
      "3                                           08B18876   \n",
      "4                                            08B260A   \n",
      "\n",
      "                               ID_ORIGINAL_NHC_BBANK  \\\n",
      "0  Unique identifier from Hospital (LP, OVE, MDA ...   \n",
      "1  Unique alphanumeric code from the hospital or ...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                                           HISTOLOGY  \\\n",
      "0                         Tumor histology/Tumor type   \n",
      "1  0=endometroid; 1=clear cells; 2=mixed; 3=other...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                                    HISTOLOGY_DETAIL  \\\n",
      "0  Tumor type based on the present cells: mixed (...   \n",
      "1         type of mixed cancer; no mixed; NA=unknown   \n",
      "2                                           No mixed   \n",
      "3                                           No mixed   \n",
      "4                                           No mixed   \n",
      "\n",
      "                                                 AGE  \\\n",
      "0  Age at diagnosis. Calculated as (date of diagn...   \n",
      "1                            2 digits; unit is YEARS   \n",
      "2                                                 50   \n",
      "3                                                 42   \n",
      "4                                                 68   \n",
      "\n",
      "                                               GRADE  \\\n",
      "0                        Tumor differentiation grade   \n",
      "1  1=well differentiated; 2=moderately differenti...   \n",
      "2                                                  2   \n",
      "3                                                  2   \n",
      "4                                                  2   \n",
      "\n",
      "                                                FIGO  \\\n",
      "0                                   FIGO tumor stage   \n",
      "1  1=IA; 2=IB; 3=IC; 4=I(NOS); 5=IIA; 6=IIB; 8=II...   \n",
      "2                                                  3   \n",
      "3                                                  3   \n",
      "4                                                  5   \n",
      "\n",
      "                                               FIGOL  \\\n",
      "0  FIGO tumor stage according to cancer spread (l...   \n",
      "1  0=localized (I,II); 1=advanced(III,IV); NA=unk...   \n",
      "2                                                  0   \n",
      "3                                                  0   \n",
      "4                                                  0   \n",
      "\n",
      "                                               FIGOa  ...  \\\n",
      "0                             Aggregated tumor stage  ...   \n",
      "1  1=IA,IB,IC,I(NOS); 2=IIA,IIB,II(NOS); 3=IIIA,I...  ...   \n",
      "2                                                  1  ...   \n",
      "3                                                  1  ...   \n",
      "4                                                  2  ...   \n",
      "\n",
      "                                   Non-MMRd_mutation  \\\n",
      "0  Indicates whether there are mutations in some ...   \n",
      "1  NO=No mutation in  PMS1, MLH3 and MSH3 genes; ...   \n",
      "2                                                 NO   \n",
      "3                                                 NO   \n",
      "4                                                 NO   \n",
      "\n",
      "                                     TILs_score_info  \\\n",
      "0  Indicates whether there is immunogenic informa...   \n",
      "1  YES=There is immunogenic information for this ...   \n",
      "2                                                YES   \n",
      "3                                                YES   \n",
      "4                                                YES   \n",
      "\n",
      "                                      TILs_CNIO_info  \\\n",
      "0  Indicates whether there is immunogenic informa...   \n",
      "1  YES_complete=There is complete immunogenic inf...   \n",
      "2                                       YES_complete   \n",
      "3                                       YES_complete   \n",
      "4                                       YES_complete   \n",
      "\n",
      "                                        TILs_ep_CNIO  \\\n",
      "0  Intraephitelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                1.0   \n",
      "3                                                2.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                          TILs_ep_LP  \\\n",
      "0  Intraepithelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                9.0   \n",
      "3                                                0.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                        TILs_ep_OTTA  \\\n",
      "0  Intraepithelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                9.0   \n",
      "3                                                0.0   \n",
      "4                                                1.0   \n",
      "\n",
      "                                        TILs_tu_CNIO  \\\n",
      "0  Intratumoral count (all TILs in tumor section)...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                2.0   \n",
      "3                                                2.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                          TILs_tu_LP  \\\n",
      "0  Intratumoral count (all TILs in tumor section)...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                2.0   \n",
      "3                                                0.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                         TILs_raw_ep  \\\n",
      "0  Intraepithelial raw counts (OTTA guidelines) o...   \n",
      "1  Number of TILs in the intraephitelial selected...   \n",
      "2                                                2.0   \n",
      "3                                                4.0   \n",
      "4                                               56.0   \n",
      "\n",
      "                                         TILs_raw_tu  \n",
      "0  Intratumoral raw counts (OTTA guidelines) of T...  \n",
      "1  Number of TILs in the intratumoral selected se...  \n",
      "2                                                6.0  \n",
      "3                                                7.0  \n",
      "4                                               78.0  \n",
      "\n",
      "[5 rows x 56 columns]\n",
      "                                             ID_CNIO  \\\n",
      "0                       Unique CNIO tumor identifier   \n",
      "1  LP: ID for samples from La Paz Hospital; OVE: ...   \n",
      "2                                                LP1   \n",
      "3                                                LP2   \n",
      "4                                                LP3   \n",
      "\n",
      "                                         ID_ORIGINAL  \\\n",
      "0  Unique identifier from the hospital to the pat...   \n",
      "1                                     Pathology code   \n",
      "2                                           02B28662   \n",
      "3                                           08B18876   \n",
      "4                                            08B260A   \n",
      "\n",
      "                               ID_ORIGINAL_NHC_BBANK  \\\n",
      "0  Unique identifier from Hospital (LP, OVE, MDA ...   \n",
      "1  Unique alphanumeric code from the hospital or ...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                                           HISTOLOGY  \\\n",
      "0                         Tumor histology/Tumor type   \n",
      "1  0=endometroid; 1=clear cells; 2=mixed; 3=other...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                                    HISTOLOGY_DETAIL  \\\n",
      "0  Tumor type based on the present cells: mixed (...   \n",
      "1         type of mixed cancer; no mixed; NA=unknown   \n",
      "2                                           No mixed   \n",
      "3                                           No mixed   \n",
      "4                                           No mixed   \n",
      "\n",
      "                                                 AGE  \\\n",
      "0  Age at diagnosis. Calculated as (date of diagn...   \n",
      "1                            2 digits; unit is YEARS   \n",
      "2                                                 50   \n",
      "3                                                 42   \n",
      "4                                                 68   \n",
      "\n",
      "                                               GRADE  \\\n",
      "0                        Tumor differentiation grade   \n",
      "1  1=well differentiated; 2=moderately differenti...   \n",
      "2                                                  2   \n",
      "3                                                  2   \n",
      "4                                                  2   \n",
      "\n",
      "                                                FIGO  \\\n",
      "0                                   FIGO tumor stage   \n",
      "1  1=IA; 2=IB; 3=IC; 4=I(NOS); 5=IIA; 6=IIB; 8=II...   \n",
      "2                                                  3   \n",
      "3                                                  3   \n",
      "4                                                  5   \n",
      "\n",
      "                                               FIGOL  \\\n",
      "0  FIGO tumor stage according to cancer spread (l...   \n",
      "1  0=localized (I,II); 1=advanced(III,IV); NA=unk...   \n",
      "2                                                  0   \n",
      "3                                                  0   \n",
      "4                                                  0   \n",
      "\n",
      "                                               FIGOa  ...  \\\n",
      "0                             Aggregated tumor stage  ...   \n",
      "1  1=IA,IB,IC,I(NOS); 2=IIA,IIB,II(NOS); 3=IIIA,I...  ...   \n",
      "2                                                  1  ...   \n",
      "3                                                  1  ...   \n",
      "4                                                  2  ...   \n",
      "\n",
      "                                   Non-MMRd_mutation  \\\n",
      "0  Indicates whether there are mutations in some ...   \n",
      "1  NO=No mutation in  PMS1, MLH3 and MSH3 genes; ...   \n",
      "2                                                 NO   \n",
      "3                                                 NO   \n",
      "4                                                 NO   \n",
      "\n",
      "                                     TILs_score_info  \\\n",
      "0  Indicates whether there is immunogenic informa...   \n",
      "1  YES=There is immunogenic information for this ...   \n",
      "2                                                YES   \n",
      "3                                                YES   \n",
      "4                                                YES   \n",
      "\n",
      "                                      TILs_CNIO_info  \\\n",
      "0  Indicates whether there is immunogenic informa...   \n",
      "1  YES_complete=There is complete immunogenic inf...   \n",
      "2                                       YES_complete   \n",
      "3                                       YES_complete   \n",
      "4                                       YES_complete   \n",
      "\n",
      "                                        TILs_ep_CNIO  \\\n",
      "0  Intraephitelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                1.0   \n",
      "3                                                2.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                          TILs_ep_LP  \\\n",
      "0  Intraepithelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                9.0   \n",
      "3                                                0.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                        TILs_ep_OTTA  \\\n",
      "0  Intraepithelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                9.0   \n",
      "3                                                0.0   \n",
      "4                                                1.0   \n",
      "\n",
      "                                        TILs_tu_CNIO  \\\n",
      "0  Intratumoral count (all TILs in tumor section)...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                2.0   \n",
      "3                                                2.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                          TILs_tu_LP  \\\n",
      "0  Intratumoral count (all TILs in tumor section)...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                2.0   \n",
      "3                                                0.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                         TILs_raw_ep  \\\n",
      "0  Intraepithelial raw counts (OTTA guidelines) o...   \n",
      "1  Number of TILs in the intraephitelial selected...   \n",
      "2                                                2.0   \n",
      "3                                                4.0   \n",
      "4                                               56.0   \n",
      "\n",
      "                                         TILs_raw_tu  \n",
      "0  Intratumoral raw counts (OTTA guidelines) of T...  \n",
      "1  Number of TILs in the intratumoral selected se...  \n",
      "2                                                6.0  \n",
      "3                                                7.0  \n",
      "4                                               78.0  \n",
      "\n",
      "[5 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "# Create rows with descriptions (index 0) and coding (index 1) now in the new file. In the order showed with info().\n",
    "# Step 1: Create dictionaries with the information\n",
    "descriptions = {\n",
    "    'ID_CNIO': 'Unique CNIO tumor identifier',\n",
    "    'ID_ORIGINAL':'Unique identifier from the hospital to the patient/sample',\n",
    "    'ID_ORIGINAL_NHC_BBANK':'Unique identifier from Hospital (LP, OVE, MDA series; NHC= number of clinical history) or Biobank (RVB series) to the patient/sample',\n",
    "    'HISTOLOGY': 'Tumor histology/Tumor type',\n",
    "    'HISTOLOGY_DETAIL':'Tumor type based on the present cells: mixed (type of mixed in that case), no mixed',\n",
    "    'AGE': 'Age at diagnosis. Calculated as (date of diagnosis – date of birth). When not provided, date of surgery was used instead of date of diagnosis (OVE series); a tag column was added to indicate this: “DATE_AGE”',\n",
    "    'GRADE':'Tumor differentiation grade',\n",
    "    'FIGO':'FIGO tumor stage',\n",
    "    'FIGOL':'FIGO tumor stage according to cancer spread (localized or advance)',\n",
    "    'FIGOa':'Aggregated tumor stage',\n",
    "    'NEOADJUVANT_TREATMENT':'Information about whether the patient received neoadjuvant treatment',\n",
    "    'ADJUVANT_TREATMENT':'Information about whether the patient received chemotherapy',\n",
    "    'TYPE_ADJUVANT':'Type of adjuvant treatment received. Carboplatino+taxol is the standard',\n",
    "    'OTHER_ADJ_TREAT':'Other type of adjuvant treatment received by the patient different from the standard, carboplatino+taxol',\n",
    "    'RESIDUAL':'Residual disease after surgery',\n",
    "    'RESIDUALa':'Aggregated categories of residual disease after surgery',\n",
    "    'OS_CNIO':'Overall survival,  TIME FROM DIAGNOSIS TO DEATH or LAST KNOWN TO BE ALIVE. Calculated by CNIO from original records as (date of death MINUS date of diagnosis)  or (date last known to be alive MINUS date of diagnosis). When not provided, date of surgery was used instead of date of diagnosis. In this last case it is indicated in the tag column DATE USED FOR OS',\n",
    "    'VITAL STATUS':'Vital status at last followup',\n",
    "    'MMR STATUS':'Information about the GERMILINE mutations in genes related to mismatch repair pathway',\n",
    "    'BRCA STATUS':'Information about GERMILINE mutatioS in BRCA1/2 genes',\n",
    "    'FAMILIAL':'Familial antecedents of colon,endometrial and/or ovarian cancer. In OVE series this information refers to possible hereditary conditions according to the clinician',\n",
    "    'DATE_AGE':'Tag column to indicate if the date used for calculating age at diagnosis was the diagnosis date or the surgery date',\n",
    "    'DATE_OS':'Tag column to indicate how overall survival was calculated, using date of diagnosis or date of surgery (OVE series)',\n",
    "    'PARTIAL DATE DIAGNOSIS':'Tag column to indicate how overall survival was calculated, using a partial or a complete date of diagnosis. If a partial date is used this suggests that the estimation of OS is not exact. A partial date is coded as XX/XX/year or XX/month/year and to make it possible the calculation these are transformed into 01/06/year or 01/month/year respectively',\n",
    "    'PARTIAL DATE DEATH_LASTv':'Tag column to indicate how overall survival was calculated, using a partial or a complete date of last visit or death. If a partial date is used this suggests that the estimation of OS is not exact. A partial date is coded as XX/XX/year or XX/month/year and to make it possible the calculation these are transformed into 01/06/year or 01/month/year respectively',\n",
    "    'Clinical_info':'Indicates whether there is clinical information available for this sample',\n",
    "    'IHC_defective_MMR_protein':'IHC defective MMR protein according the results of immunohistochemistry tests. IHC tests the expression of MLH1, MSH2, MSH6 and PMS2 in tumour tissue sections. MMRp tumours express nuclear positivity of these 4 proteins, whereas MMRd tumours generally exhibit loss of expression of some of these proteins. Defective protein according to staining pattern of MLH1, PMS2, MSH2 & MSH6',\n",
    "    'IHC pattern':'It refers to Abnormal or Standard IHC pattern for the missing proteins in the IHC',\n",
    "    'IHC_MMR_class':'MMR classification based on IHC results of immunohistochemistry tests. IHC tests the expression of MLH1, MSH2, MSH6 and PMS2 in tumour tissue sections. MMRp tumours show nuclear positivity of these 4 proteins, whereas MMRd tumours generally exhibit loss of expression of some of these proteins',\n",
    "    'PCR':'Results of PCR tests. Classic PCR test for MSI detection examines five MS regions in the genes NR21, NR22, NR24, BAT25 and BAT26. MSS tumours show no degree of instability across these 5 markers, whereas instability at just one marker is defined as MSI-Low. Instability at two or more of these markers confers a MSI-High diagnosis',\n",
    "    'Evidence score for MMRd':'Evidence score for MMR deficiency based on data availability and agreement of standard techniques',\n",
    "    'IHCd':'Results of immunohistochemistry tests used in Diego’s work (refers to ‘IHC_lab’ data)',\n",
    "    'PCRd':'PCR results used in Diego’s work (refers to ‘PCR_lab’ data)',\n",
    "    'QC_filters':'Indicates if this sample has passed NGS QC filters: more than 2 million reads, certain coverage and not artefacts in variant calling',\n",
    "    'Genomic_info':'Indicates whether there is genomic information available for ALL of SOME of the features tested in this sample',\n",
    "    'MMR_final_status':'MMR classification according to gold standard techniques (IHC and PCR)',\n",
    "    'CLASSd':'Classification of the MMR system for this sample according to Diego’s work',\n",
    "    'MSI_sensor2':'Microsatellite instability quantified as a numeric score using MSIsensor2 algorithm. A higher score indicates a higher microsatellite instability in that tumour.The MSI score refers to the number of MSI sites/all valid sites',\n",
    "    'TMB': 'Total number of non synonymous coding mutations per megabase (Mb) of a tumour genome',\n",
    "    'ID2':'Indel (Insertion-deletion) mutational signature. Biomarker linked to MMRd. It is predominantly composed of deletions of thymine at long (≥5) thymine mononucleotide repeats',\n",
    "    'ID7':'Indel (Insertion-deletion) mutational signature. Biomarker linked to MMRd. Mutational signature indicative of large number of indels',\n",
    "    'SUM ID2+ID7':'Sum of the indel (Insertion-deletion) mutational signatures ID2 and ID7',\n",
    "    'CNV': 'Total number of copy number alterations (events) per tumor sample',\n",
    "    '%genome_altered':'Percentage of the genome mutated/altered from the total',\n",
    "    'Pol_mutation':'Indicates whether there are mutations in some of polymerase genes. Here we refers to the genes POLE or POLD1. POLE encodes the catalytic subunit of DNA polymerase epsilon while POLD1 encodes the catalytic and proofreading subunit of DNA polymerase-delta',\n",
    "    'MMRd_mutation':'Indicates whether there are mutations in some of the genes involved in MMR pathway. Here we refers to the genes MLH1, MSH2, MSH6 and PMS2',\n",
    "    'Non-MMRd_mutation':'Indicates whether there are mutations in some of the genes involved in MMR pathway. Here we refers to the genes  PMS1, MLH3 and MSH3',\n",
    "    'TILs_score_info':'Indicates whether there is immunogenic information available for this sample. It refers to scores information. ‘TILs’ term refers to cytotoxic CD8+ tumour-infiltrating lymphocytes',\n",
    "    'TILs_CNIO_info':'Indicates whether there is immunogenic information (CD8+ tumour-infiltrating lymphocytes) available for this sample coming from CNIO and if this information is complete or incomplete and in what sense',\n",
    "    'TILs_ep_CNIO':'Intraephitelial count (OTTA guidelines) of TILs made by Eduardo Caleiras at CNIO',\n",
    "    'TILs_ep_LP':'Intraepithelial count (OTTA guidelines) of TILs made by Berjón at La Paz Hospital',\n",
    "    'TILs_ep_OTTA':'Intraepithelial count (OTTA guidelines) of TILs made by OTTA in Canada',\n",
    "    'TILs_tu_CNIO':'Intratumoral count (all TILs in tumor section) made at CNIO',\n",
    "    'TILs_tu_LP':'Intratumoral count (all TILs in tumor section) made by Berjón at La Paz Hospital',\n",
    "    'TILs_raw_ep':'Intraepithelial raw counts (OTTA guidelines) of TILs made by Eduardo Caleiras at CNIO.',\n",
    "    'TILs_raw_tu':'Intratumoral raw counts (OTTA guidelines) of TILs made by Eduardo Caleiras at CNIO'\n",
    "}\n",
    "subheaders = ['LP: ID for samples from La Paz Hospital; OVE: ID for samples from Virgen del Rocio Hospital; MDA:ID for samples from MD Anderson Cancer Center  Hospital ; RVB: ID for samples from Red Valenciana de Biobancos', \n",
    "              'Pathology code','Unique alphanumeric code from the hospital or biobank','0=endometroid; 1=clear cells; 2=mixed; 3=others; NA=unknown', \n",
    "             'type of mixed cancer; no mixed; NA=unknown','2 digits; unit is YEARS',\n",
    "             '1=well differentiated; 2=moderately differentiated; 3=poorly differentiated; NA=unknown',\n",
    "             '1=IA; 2=IB; 3=IC; 4=I(NOS); 5=IIA; 6=IIB; 8=II(NOS); 9=IIIA; 10=IIIB; 11=IIIC; 12=III(NOS); 13=IV; NA=unknown',\n",
    "             '0=localized (I,II); 1=advanced(III,IV); NA=unknown','1=IA,IB,IC,I(NOS); 2=IIA,IIB,II(NOS); 3=IIIA,IIIB,IIIC,III(NOS); 4=IV; NA=unknown',\n",
    "             '0=NO; 1=YES; NA=unknown','0=NO; 1=YES; NA=unknown',\n",
    "             '0=Carbo-paclitaxel; 1=Cis-paclitaxel; 2=Carbo-monotherapy; 3=Carbo-Taxol-Beva; 4=other; NA=unknown',\n",
    "             'Drugs used as adjuvant treatment',\n",
    "             '0=no macroscopic disease; 1=macroscopic disease<=1; 2=macroscopic disease>1; 3=macroscopic disease,unknown size; NA=unknown',\n",
    "             '0=No residual disease; 1=Yes residual disease; NA=unknown','Unit is DAYS',\n",
    "             '0=alive; 1=dead; NA=unknown','0=not studied; 1=studied,without mutation; 2=mutation in MMR genes; 3=other genes mutated; NA=unknown',\n",
    "             '0=not studied; 1=studied,without mutation; 2=mutation in BRCA1/2; 3=other genes mutated; NA=unknown','0=NO; 1=YES; NA=unknown',\n",
    "             'D=diagnosis date used for calculating age; S=surgery date used for calculating age; NA=unknown',\n",
    "             'D=diagnosis date used for OS; S=surgery date used for OS; LD=Lacking diagnosis/surgery/death_lastv date; NA=unknown',\n",
    "             'YES=partial date xx/xx/y or xx/m/y; NO=Complete date or NA','YES=partial date xx/xx/y or xx/m/y; NO=Complete date or NA',\n",
    "             'YES=There is clinical information for this sample; NO=There isn’t clinical information for this sample',\n",
    "             'Defective protein according to staining pattern of MLH1, PMS2, MSH2 & MSH6: MLH1=MLH1 & PMS2 negative staining; MSH2=MSH2 & MSH6 negative staining; PMS2=PMS2 negat staining;  MSH6=MSH6 negative staining; NO=All positive; NA=Not Available',\n",
    "             'Standard= see column \"IHC_defective_MMR_protein\"; Abnormal=negat staining of MLH1,PMS2,MSH6 or MSH2 in a combination which is not the standard; NO=All positive; NA=Not Available',\n",
    "             'MMR proficient=All four MLH1, PMS2, MSH2 & MSH6 proteins showed positive staining; MMRd, MMR deficient=One/two of the MLH1, PMS2, MSH2 & MSH6 showed negative staining;  NA=Not Available',\n",
    "             'MSS=Microsatellite stability-no degree of instability across the 5 tested markers; MSI-Low=Low microsatellite instability-instability at one tested marker; MSI-High=High microsatellite instability-instability at two or more of the tested markers',\n",
    "             '1=IHC&PCR indicates MMRd unambiguously; 2= One MMRd unambiguous, the other MMRd ambiguous or NA;  3= One MMRd unambiguous, the other MMRp or NA; 4= One MMRd ambiguous',\n",
    "             'MLH1, MSH2, MSH6 and PMS2: indicates the loss of expression of that specific protein; UNKNOWN: tested, but unknown',\n",
    "             'MSS= Microsatellite stability-no degree of instability across the 5 tested markers; MSI-Low=Low microsatellite instability-instability at one tested marker; MSI-High=High microsatellite instability-instability at two or more of the tested markers',\n",
    "             'YES=Sample passes QC filters; NO=Sample doesn’t pass QC filters',\n",
    "             'YES=There is genomic information for this sample; NO=There isn’t genomic information for this sample',\n",
    "             'MMRp=MMR proficient sample, MMRd=MMR deficient sample. Only samples with MMRd evidence score equal or greater than 3 were classified as MMRd',\n",
    "             'MMRp:mismatch repair proficient sample; MMRd:mismatch repair deficient sample',\n",
    "             'Percentage','Number of non synonymous coding mutations/Mb tumour genome',\n",
    "             'Percentage of observed mutations in the tumour sample that can be attributed to ID2 mutational signature',\n",
    "             'Percentage of observed mutations in the tumour sample that can be attributed to ID7 mutational signature',\n",
    "             'Percentage of observed mutations in the tumour sample that can be attributed to ID2+ID7 mutational signatures',\n",
    "             'Number of copy number alterations/events.One event is considered when a DNA segment has more or fewer than 2 copies (the normal value)',\n",
    "             'Percentage','NO=No mutation in polymerase genes; POLE:POLE gene is mutated (in one copy of the genome); POLD1:POLD1 gene is mutated (in one copy of the genome); POLE&POLE:Both POLE copies in the genome are mutated',\n",
    "             'NO=No mutation in MMR genes; MLH1,MSH2,MSH6 and PMS2:indicates a mutation in that specific MMR gene.Sometimes there are more than one gene mutated or there are two copies of the same gene mutated (&)',\n",
    "             'NO=No mutation in  PMS1, MLH3 and MSH3 genes; PMS1, MLH3 and MSH3:indicates a mutation in that specific gene',\n",
    "             'YES=There is immunogenic information for this sample; NO=There isn’t immunogenic information for this sample',\n",
    "             'YES_complete=There is complete immunogenic information for this sample, intraephitelial and intratumoral data; YES_only_ep= There is only information about the number of intraepithelial lymphocytes ; YES_only_tu: There is only information about the number of intratumoral lymphocytes; NO=There isn’t immunogenic information for this sample',\n",
    "             '0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs; 9:not evaluable','0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs; 9:not evaluable',\n",
    "             '0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs; 9:not evaluable','0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs; 9:not evaluable',\n",
    "             '0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs; 9:not evaluable',\n",
    "              'Number of TILs in the intraephitelial selected section of TMA','Number of TILs in the intratumoral selected section of TMA',\n",
    "             ]\n",
    "\n",
    "# Step 2: Convert 'descriptions' and 'subheaders-coding' into a DataFrame\n",
    "# Align descriptions and coding with the columns in 'new_df'\n",
    "descriptions_row = pd.DataFrame([descriptions], columns=final_df.columns)\n",
    "coding_row = pd.DataFrame([subheaders], columns=final_df.columns)\n",
    "\n",
    "# Step 3: Concatenate the new rows with the original DataFrame\n",
    "new_df_with_annotations = pd.concat([descriptions_row, coding_row, final_df], ignore_index=True)\n",
    "\n",
    "# Step 4: Display or save the new DataFrame\n",
    "print(new_df_with_annotations.head(5))  # Show the first 5 rows\n",
    "# Save the DataFrame to Excel if needed\n",
    "new_df_with_annotations.to_excel('/home/vant/TFM/Final_db/Combined_data_all_cohorts_annotated.xlsx', index=False) # 330 samples: some with only tils,genomic or clinical,\n",
    "# with 2 of them o with the 3 of them\n",
    "\n",
    "# Final file with samples that have the 3 types of data at the same time\n",
    "descriptions_row1 = pd.DataFrame([descriptions], columns=new_df.columns)\n",
    "coding_row1 = pd.DataFrame([subheaders], columns=new_df.columns)\n",
    "\n",
    "# Step 3: Concatenate the new rows with the original DataFrame\n",
    "new_df_with_annotations1 = pd.concat([descriptions_row1, coding_row1, new_df], ignore_index=True)\n",
    "\n",
    "# Step 4: Display or save the new DataFrame\n",
    "print(new_df_with_annotations1.head(5))\n",
    "new_df_with_annotations1.to_excel(\"/home/vant/TFM/Final_db/Samples_alltypedata_annotated.xlsx\",index=False) # Samples with the 3 types of datap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c9f0b577-937e-4b94-8131-cf5331e02b1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             ID_CNIO  \\\n",
      "0                       Unique CNIO tumor identifier   \n",
      "1  LP: ID for samples from La Paz Hospital; OVE: ...   \n",
      "2                                                LP1   \n",
      "3                                                LP2   \n",
      "4                                                LP3   \n",
      "\n",
      "                                         ID_ORIGINAL  \\\n",
      "0  Unique identifier from the hospital to the pat...   \n",
      "1                                     Pathology code   \n",
      "2                                           02B28662   \n",
      "3                                           08B18876   \n",
      "4                                            08B260A   \n",
      "\n",
      "                               ID_ORIGINAL_NHC_BBANK  \\\n",
      "0  Unique identifier from Hospital (LP, OVE, MDA ...   \n",
      "1  Unique alphanumeric code from the hospital or ...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                                           HISTOLOGY  \\\n",
      "0                         Tumor histology/Tumor type   \n",
      "1  0=endometroid; 1=clear cells; 2=mixed; 3=other...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                                    HISTOLOGY_DETAIL  \\\n",
      "0  Tumor type based on the present cells: mixed (...   \n",
      "1         type of mixed cancer; no mixed; NA=unknown   \n",
      "2                                           No mixed   \n",
      "3                                           No mixed   \n",
      "4                                           No mixed   \n",
      "\n",
      "                                                 AGE  \\\n",
      "0  Age at diagnosis. Calculated as (date of diagn...   \n",
      "1                            2 digits; unit is YEARS   \n",
      "2                                                 50   \n",
      "3                                                 42   \n",
      "4                                                 68   \n",
      "\n",
      "                                               GRADE  \\\n",
      "0                        Tumor differentiation grade   \n",
      "1  1=well differentiated; 2=moderately differenti...   \n",
      "2                                                  2   \n",
      "3                                                  2   \n",
      "4                                                  2   \n",
      "\n",
      "                                                FIGO  \\\n",
      "0                                   FIGO tumor stage   \n",
      "1  1=IA; 2=IB; 3=IC; 4=I(NOS); 5=IIA; 6=IIB; 8=II...   \n",
      "2                                                  3   \n",
      "3                                                  3   \n",
      "4                                                  5   \n",
      "\n",
      "                                               FIGOL  \\\n",
      "0  FIGO tumor stage according to cancer spread (l...   \n",
      "1  0=localized (I,II); 1=advanced(III,IV); NA=unk...   \n",
      "2                                                  0   \n",
      "3                                                  0   \n",
      "4                                                  0   \n",
      "\n",
      "                                               FIGOa  ...  \\\n",
      "0                             Aggregated tumor stage  ...   \n",
      "1  1=IA,IB,IC,I(NOS); 2=IIA,IIB,II(NOS); 3=IIIA,I...  ...   \n",
      "2                                                  1  ...   \n",
      "3                                                  1  ...   \n",
      "4                                                  2  ...   \n",
      "\n",
      "                                           RESIDUALa  \\\n",
      "0  Aggregated categories of residual disease afte...   \n",
      "1  0=No residual disease; 1=Yes residual disease;...   \n",
      "2                                                  0   \n",
      "3                                                  0   \n",
      "4                                                  0   \n",
      "\n",
      "                                             OS_CNIO  \\\n",
      "0  Overall survival,  TIME FROM DIAGNOSIS TO DEAT...   \n",
      "1                                       Unit is DAYS   \n",
      "2                                               5568   \n",
      "3                                               5163   \n",
      "4                                               3047   \n",
      "\n",
      "                    VITAL STATUS  \\\n",
      "0  Vital status at last followup   \n",
      "1    0=alive; 1=dead; NA=unknown   \n",
      "2                              0   \n",
      "3                              0   \n",
      "4                              0   \n",
      "\n",
      "                                          MMR STATUS  \\\n",
      "0  Information about the GERMILINE mutations in g...   \n",
      "1  0=not studied; 1=studied,without mutation; 2=m...   \n",
      "2                                                  0   \n",
      "3                                                  0   \n",
      "4                                                  0   \n",
      "\n",
      "                                         BRCA STATUS  \\\n",
      "0  Information about GERMILINE mutatioS in BRCA1/...   \n",
      "1  0=not studied; 1=studied,without mutation; 2=m...   \n",
      "2                                                  0   \n",
      "3                                                  0   \n",
      "4                                                  0   \n",
      "\n",
      "                                            FAMILIAL  \\\n",
      "0  Familial antecedents of colon,endometrial and/...   \n",
      "1                            0=NO; 1=YES; NA=unknown   \n",
      "2                                                  0   \n",
      "3                                                  0   \n",
      "4                                                  1   \n",
      "\n",
      "                                            DATE_AGE  \\\n",
      "0  Tag column to indicate if the date used for ca...   \n",
      "1  D=diagnosis date used for calculating age; S=s...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                                             DATE_OS  \\\n",
      "0  Tag column to indicate how overall survival wa...   \n",
      "1  D=diagnosis date used for OS; S=surgery date u...   \n",
      "2                                                  D   \n",
      "3                                                  D   \n",
      "4                                                  D   \n",
      "\n",
      "                              PARTIAL DATE DIAGNOSIS  \\\n",
      "0  Tag column to indicate how overall survival wa...   \n",
      "1  YES=partial date xx/xx/y or xx/m/y; NO=Complet...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                            PARTIAL DATE DEATH_LASTv  \n",
      "0  Tag column to indicate how overall survival wa...  \n",
      "1  YES=partial date xx/xx/y or xx/m/y; NO=Complet...  \n",
      "2                                                NaN  \n",
      "3                                                NaN  \n",
      "4                                                NaN  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "# Create the clinical database\n",
    "# Create a new DataFrame with rows with clinical information and columns from index 0 to 25 to extract only the clinical database\n",
    "filtered_clinical=new_df_with_annotations[new_df_with_annotations['Clinical_info']=='YES']\n",
    "# Select also rows with índexes 0 and 1\n",
    "filtered_clinical = pd.concat([new_df_with_annotations.loc[[0, 1]], filtered_clinical])\n",
    "new_df_clinical = filtered_clinical.iloc[:, :25]\n",
    "# Display the first few rows of the new DataFrame to verify\n",
    "print(new_df_clinical.head())\n",
    "new_df_clinical.to_excel(\"/home/vant/TFM/Final_db/Definitive_clinical_database.xlsx\", index=False) #309 samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "98250aef-5cc3-4f48-a705-d0b035a4f5ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             ID_CNIO  \\\n",
      "0                       Unique CNIO tumor identifier   \n",
      "1  LP: ID for samples from La Paz Hospital; OVE: ...   \n",
      "2                                                LP1   \n",
      "3                                                LP2   \n",
      "4                                                LP3   \n",
      "\n",
      "                                         ID_ORIGINAL  \\\n",
      "0  Unique identifier from the hospital to the pat...   \n",
      "1                                     Pathology code   \n",
      "2                                           02B28662   \n",
      "3                                           08B18876   \n",
      "4                                            08B260A   \n",
      "\n",
      "                               ID_ORIGINAL_NHC_BBANK  \\\n",
      "0  Unique identifier from Hospital (LP, OVE, MDA ...   \n",
      "1  Unique alphanumeric code from the hospital or ...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                                           HISTOLOGY  \\\n",
      "0                         Tumor histology/Tumor type   \n",
      "1  0=endometroid; 1=clear cells; 2=mixed; 3=other...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                           IHC_defective_MMR_protein  \\\n",
      "0  IHC defective MMR protein according the result...   \n",
      "1  Defective protein according to staining patter...   \n",
      "2                                                 NO   \n",
      "3                                                 NO   \n",
      "4                                                 NO   \n",
      "\n",
      "                                         IHC pattern  \\\n",
      "0  It refers to Abnormal or Standard IHC pattern ...   \n",
      "1  Standard= see column \"IHC_defective_MMR_protei...   \n",
      "2                                                 NO   \n",
      "3                                                 NO   \n",
      "4                                                 NO   \n",
      "\n",
      "                                       IHC_MMR_class  \\\n",
      "0  MMR classification based on IHC results of imm...   \n",
      "1  MMR proficient=All four MLH1, PMS2, MSH2 & MSH...   \n",
      "2                                               MMRp   \n",
      "3                                               MMRp   \n",
      "4                                               MMRp   \n",
      "\n",
      "                                                 PCR  \\\n",
      "0  Results of PCR tests. Classic PCR test for MSI...   \n",
      "1  MSS=Microsatellite stability-no degree of inst...   \n",
      "2                                                MSS   \n",
      "3                                                MSS   \n",
      "4                                                MSS   \n",
      "\n",
      "                             Evidence score for MMRd  \\\n",
      "0  Evidence score for MMR deficiency based on dat...   \n",
      "1  1=IHC&PCR indicates MMRd unambiguously; 2= One...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                                                IHCd  ...  \\\n",
      "0  Results of immunohistochemistry tests used in ...  ...   \n",
      "1  MLH1, MSH2, MSH6 and PMS2: indicates the loss ...  ...   \n",
      "2                                                 NO  ...   \n",
      "3                                                 NO  ...   \n",
      "4                                                 NO  ...   \n",
      "\n",
      "                                         MSI_sensor2  \\\n",
      "0  Microsatellite instability quantified as a num...   \n",
      "1                                         Percentage   \n",
      "2                                               4.55   \n",
      "3                                               9.43   \n",
      "4                                               6.38   \n",
      "\n",
      "                                                 TMB  \\\n",
      "0  Total number of non synonymous coding mutation...   \n",
      "1  Number of non synonymous coding mutations/Mb t...   \n",
      "2                                          17.003493   \n",
      "3                                           7.557108   \n",
      "4                                           7.557108   \n",
      "\n",
      "                                                 ID2  \\\n",
      "0  Indel (Insertion-deletion) mutational signatur...   \n",
      "1  Percentage of observed mutations in the tumour...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                                                 ID7  \\\n",
      "0  Indel (Insertion-deletion) mutational signatur...   \n",
      "1  Percentage of observed mutations in the tumour...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                                         SUM ID2+ID7  \\\n",
      "0  Sum of the indel (Insertion-deletion) mutation...   \n",
      "1  Percentage of observed mutations in the tumour...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                                                 CNV  \\\n",
      "0  Total number of copy number alterations (event...   \n",
      "1  Number of copy number alterations/events.One e...   \n",
      "2                                                1.0   \n",
      "3                                                0.0   \n",
      "4                                                4.0   \n",
      "\n",
      "                                     %genome_altered  \\\n",
      "0  Percentage of the genome mutated/altered from ...   \n",
      "1                                         Percentage   \n",
      "2                                           3.014842   \n",
      "3                                                0.0   \n",
      "4                                           9.044527   \n",
      "\n",
      "                                        Pol_mutation  \\\n",
      "0  Indicates whether there are mutations in some ...   \n",
      "1  NO=No mutation in polymerase genes; POLE:POLE ...   \n",
      "2                                                 NO   \n",
      "3                                                 NO   \n",
      "4                                                 NO   \n",
      "\n",
      "                                       MMRd_mutation  \\\n",
      "0  Indicates whether there are mutations in some ...   \n",
      "1  NO=No mutation in MMR genes; MLH1,MSH2,MSH6 an...   \n",
      "2                                                 NO   \n",
      "3                                                 NO   \n",
      "4                                                 NO   \n",
      "\n",
      "                                   Non-MMRd_mutation  \n",
      "0  Indicates whether there are mutations in some ...  \n",
      "1  NO=No mutation in  PMS1, MLH3 and MSH3 genes; ...  \n",
      "2                                                 NO  \n",
      "3                                                 NO  \n",
      "4                                                 NO  \n",
      "\n",
      "[5 rows x 25 columns]\n",
      "                                             ID_CNIO  \\\n",
      "0                       Unique CNIO tumor identifier   \n",
      "1  LP: ID for samples from La Paz Hospital; OVE: ...   \n",
      "2                                                LP1   \n",
      "3                                                LP2   \n",
      "4                                                LP3   \n",
      "\n",
      "                                         ID_ORIGINAL  \\\n",
      "0  Unique identifier from the hospital to the pat...   \n",
      "1                                     Pathology code   \n",
      "2                                           02B28662   \n",
      "3                                           08B18876   \n",
      "4                                            08B260A   \n",
      "\n",
      "                               ID_ORIGINAL_NHC_BBANK  \\\n",
      "0  Unique identifier from Hospital (LP, OVE, MDA ...   \n",
      "1  Unique alphanumeric code from the hospital or ...   \n",
      "2                                                NaN   \n",
      "3                                                NaN   \n",
      "4                                                NaN   \n",
      "\n",
      "                                           HISTOLOGY  \\\n",
      "0                         Tumor histology/Tumor type   \n",
      "1  0=endometroid; 1=clear cells; 2=mixed; 3=other...   \n",
      "2                                                0.0   \n",
      "3                                                0.0   \n",
      "4                                                0.0   \n",
      "\n",
      "                                     TILs_score_info  \\\n",
      "0  Indicates whether there is immunogenic informa...   \n",
      "1  YES=There is immunogenic information for this ...   \n",
      "2                                                YES   \n",
      "3                                                YES   \n",
      "4                                                YES   \n",
      "\n",
      "                                      TILs_CNIO_info  \\\n",
      "0  Indicates whether there is immunogenic informa...   \n",
      "1  YES_complete=There is complete immunogenic inf...   \n",
      "2                                       YES_complete   \n",
      "3                                       YES_complete   \n",
      "4                                       YES_complete   \n",
      "\n",
      "                                        TILs_ep_CNIO  \\\n",
      "0  Intraephitelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                1.0   \n",
      "3                                                2.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                          TILs_ep_LP  \\\n",
      "0  Intraepithelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                9.0   \n",
      "3                                                0.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                        TILs_ep_OTTA  \\\n",
      "0  Intraepithelial count (OTTA guidelines) of TIL...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                9.0   \n",
      "3                                                0.0   \n",
      "4                                                1.0   \n",
      "\n",
      "                                        TILs_tu_CNIO  \\\n",
      "0  Intratumoral count (all TILs in tumor section)...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                2.0   \n",
      "3                                                2.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                          TILs_tu_LP  \\\n",
      "0  Intratumoral count (all TILs in tumor section)...   \n",
      "1  0:no TILs; 1:1-2 TILs; 2:3-19 TILS; 3:>20 TILs...   \n",
      "2                                                2.0   \n",
      "3                                                0.0   \n",
      "4                                                3.0   \n",
      "\n",
      "                                         TILs_raw_ep  \\\n",
      "0  Intraepithelial raw counts (OTTA guidelines) o...   \n",
      "1  Number of TILs in the intraephitelial selected...   \n",
      "2                                                2.0   \n",
      "3                                                4.0   \n",
      "4                                               56.0   \n",
      "\n",
      "                                         TILs_raw_tu  \n",
      "0  Intratumoral raw counts (OTTA guidelines) of T...  \n",
      "1  Number of TILs in the intratumoral selected se...  \n",
      "2                                                6.0  \n",
      "3                                                7.0  \n",
      "4                                               78.0  \n"
     ]
    }
   ],
   "source": [
    "### Create the genomic database\n",
    "# List of specific column names you want to include in the new DataFrame\n",
    "columns_to_include = ['ID_CNIO', 'ID_ORIGINAL', 'ID_ORIGINAL_NHC_BBANK', 'HISTOLOGY','IHC_defective_MMR_protein','IHC pattern','IHC_MMR_class','PCR','Evidence score for MMRd','IHCd','PCRd','QC_filters','Genomic_info',\n",
    "                     'MMR_final_status','CLASSd','MSI_sensor2','TMB','ID2','ID7','SUM ID2+ID7','CNV','%genome_altered','Pol_mutation','MMRd_mutation',\n",
    "                      'Non-MMRd_mutation']\n",
    "### Create a TILs database\n",
    "columns_to_include2 = ['ID_CNIO', 'ID_ORIGINAL', 'ID_ORIGINAL_NHC_BBANK', 'HISTOLOGY','TILs_score_info','TILs_CNIO_info','TILs_ep_CNIO','TILs_ep_LP',\n",
    "                       'TILs_ep_OTTA','TILs_tu_CNIO','TILs_tu_LP','TILs_raw_ep','TILs_raw_tu']\n",
    "# Create a new DataFrame with only those specific columns\n",
    "filtered_genomic=new_df_with_annotations[new_df_with_annotations['Genomic_info']=='YES']\n",
    "# Select also rows with índexes 0 and 1\n",
    "filtered_genomic = pd.concat([new_df_with_annotations.loc[[0, 1]], filtered_genomic])\n",
    "new_df_genomic = filtered_genomic[columns_to_include]\n",
    "filtered_tils=new_df_with_annotations[new_df_with_annotations['TILs_score_info']=='YES']\n",
    "# Select also rows with índexes 0 and 1\n",
    "filtered_tils = pd.concat([new_df_with_annotations.loc[[0, 1]], filtered_tils])\n",
    "new_df_tils = filtered_tils[columns_to_include2]\n",
    "# Display the first few rows of the new DataFrame to verify\n",
    "print(new_df_genomic.head())\n",
    "print(new_df_tils.head())\n",
    "new_df_genomic.to_excel(\"/home/vant/TFM/Final_db/Definitive_genomic_database.xlsx\", index=False) # 173 samples\n",
    "new_df_tils.to_excel(\"/home/vant/TFM/Final_db/Definitive_TILs_database.xlsx\", index=False) # 232 samples / 195 if we only select the CNIO counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0d4f4a65-358b-425c-9c63-458c294e59dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with genomic info are:175\n"
     ]
    }
   ],
   "source": [
    "# Create the complete database with samples with genomic data\n",
    "# Filter rows in which 'Genomic_info' is 'YES'\n",
    "genomic_yes = new_df_with_annotations[new_df_with_annotations['Genomic_info'] == 'YES']\n",
    "# Select also rows with índexes 0 and 1\n",
    "genomic_yes = pd.concat([new_df_with_annotations.loc[[0, 1]], genomic_yes])\n",
    "count_genomic = genomic_yes.shape[0]\n",
    "print(f\"Number of samples with genomic info are:{count_genomic}\") # 173+2 rows (description + coding rows)\n",
    "genomic_yes.to_excel(\"/home/vant/TFM/Final_db/Samples_Genomic_definitive.xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "56cfcb58-243e-4801-8b08-c6e3e1a52877",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with clinical info are:311\n"
     ]
    }
   ],
   "source": [
    "# Create the complete database with samples with clinical data\n",
    "# Filter rows in which 'Clinical_info' is 'YES'\n",
    "clinical_yes = new_df_with_annotations[new_df_with_annotations['Clinical_info'] == 'YES']\n",
    "# Select also rows with índexes 0 and 1\n",
    "clinical_yes = pd.concat([new_df_with_annotations.loc[[0, 1]], clinical_yes])\n",
    "count_clinical = clinical_yes.shape[0]\n",
    "print(f\"Number of samples with clinical info are:{count_clinical}\")\n",
    "clinical_yes.to_excel(\"/home/vant/TFM/Final_db/Samples_Clinical_definitive.xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "43e5f0a4-b26d-4e99-9ee1-e202e5dc0017",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples with TILs info are:234\n"
     ]
    }
   ],
   "source": [
    "# Create the complete database with samples with TILs data\n",
    "# Filter rows in which 'TILs_info' is 'YES'\n",
    "tils_yes = new_df_with_annotations[new_df_with_annotations['TILs_score_info'] == 'YES']\n",
    "# Select also rows with índexes 0 and 1\n",
    "tils_yes = pd.concat([new_df_with_annotations.loc[[0, 1]], tils_yes])\n",
    "count_tils = tils_yes.shape[0]\n",
    "print(f\"Number of samples with TILs info are:{count_tils}\")\n",
    "tils_yes.to_excel(\"/home/vant/TFM/Final_db/Samples_TILs_definitive.xlsx\",index=False) # 233 in Alice's original files, 232 here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "158e6761-82c5-4d49-9390-a26ed31eba84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   ID_CNIO  ID_ORIGINAL ID_ORIGINAL_NHC_BBANK  HISTOLOGY HISTOLOGY_DETAIL AGE  \\\n",
      "0      LP1     02B28662                   NaN        0.0         No mixed  50   \n",
      "1      LP2     08B18876                   NaN        0.0         No mixed  42   \n",
      "2      LP3      08B260A                   NaN        0.0         No mixed  68   \n",
      "3      LP4     11B18858                   NaN        0.0         No mixed  50   \n",
      "4      LP5      99B6752                   NaN        0.0         No mixed  82   \n",
      "5      LP6      00B9877                   NaN        0.0         No mixed  62   \n",
      "6      LP7     12B25893                   NaN        0.0         No mixed  45   \n",
      "7      LP8     10B25690                   NaN        0.0         No mixed  48   \n",
      "8      LP9     08B10682                   NaN        0.0         No mixed  51   \n",
      "9     LP10      00B6886                   NaN        0.0         No mixed  56   \n",
      "10    LP11     03B2020A                   NaN        0.0         No mixed  34   \n",
      "11    LP12     08B27334                   NaN        0.0         No mixed  54   \n",
      "12    LP13      09B7449                   NaN        0.0         No mixed  73   \n",
      "13    LP14     12B14655                   NaN        0.0         No mixed  52   \n",
      "14    LP15     00B18539                   NaN        0.0         No mixed  82   \n",
      "15    LP16     98B26187                   NaN        0.0         No mixed  40   \n",
      "16    LP17      98B2092                   NaN        0.0         No mixed  52   \n",
      "17    LP18     02B17242                   NaN        0.0         No mixed  60   \n",
      "18    LP19    05B11475A                   NaN        0.0         No mixed  68   \n",
      "19    LP20  06B15465 I?                   NaN        0.0         No mixed  59   \n",
      "20    LP21      10B6659                   NaN        0.0         No mixed  51   \n",
      "21    LP22      12B7328                   NaN        0.0         No mixed  38   \n",
      "22    LP23          NaN                   NaN        0.0         No mixed  45   \n",
      "23    LP24     98B02808                   NaN        0.0         No mixed  57   \n",
      "24    LP25     10B10096                   NaN        0.0         No mixed  39   \n",
      "\n",
      "   GRADE FIGO FIGOL FIGOa  ... Non-MMRd_mutation TILs_score_info  \\\n",
      "0      2    3     0     1  ...                NO             YES   \n",
      "1      2    3     0     1  ...                NO             YES   \n",
      "2      2    5     0     2  ...                NO             YES   \n",
      "3      1    6     0     2  ...                NO             YES   \n",
      "4      1    1     0     1  ...                NO             YES   \n",
      "5      3   11     1     3  ...               NaN             YES   \n",
      "6      1    3     0     1  ...               NaN             YES   \n",
      "7      1    2     0     1  ...                NO             YES   \n",
      "8      1    2     0     1  ...                NO             YES   \n",
      "9      1    3     0     1  ...               NaN             YES   \n",
      "10     2    3     0     1  ...                NO             YES   \n",
      "11     1    1     0     1  ...                NO             YES   \n",
      "12     2   13     1     4  ...                NO             YES   \n",
      "13     1    3     0     1  ...                NO             YES   \n",
      "14   NaN    1     0     1  ...                NO             YES   \n",
      "15     1    1     0     1  ...                NO             YES   \n",
      "16     3    1     0     1  ...                NO             YES   \n",
      "17     1    1     0     1  ...              MSH3             YES   \n",
      "18     2    9     1     3  ...               NaN             YES   \n",
      "19     1    1     0     1  ...                NO             YES   \n",
      "20     2    3     0     1  ...                NO             YES   \n",
      "21     2   11     1     3  ...                NO             YES   \n",
      "22     1    3     0     1  ...               NaN              NO   \n",
      "23     2    8     0     2  ...                NO             YES   \n",
      "24     1   12     1     3  ...               NaN             YES   \n",
      "\n",
      "   TILs_CNIO_info TILs_ep_CNIO TILs_ep_LP TILs_ep_OTTA TILs_tu_CNIO  \\\n",
      "0    YES_complete          1.0        9.0          9.0          2.0   \n",
      "1    YES_complete          2.0        0.0          0.0          2.0   \n",
      "2    YES_complete          3.0        3.0          1.0          3.0   \n",
      "3    YES_complete          2.0        3.0          2.0          2.0   \n",
      "4              NO          NaN        1.0          0.0          NaN   \n",
      "5    YES_complete          2.0        NaN          NaN          3.0   \n",
      "6    YES_complete          1.0        3.0          0.0          2.0   \n",
      "7    YES_complete          1.0        2.0          1.0          2.0   \n",
      "8    YES_complete          2.0        2.0          0.0          3.0   \n",
      "9              NO          NaN        0.0          0.0          NaN   \n",
      "10   YES_complete          0.0        NaN          NaN          3.0   \n",
      "11   YES_complete          0.0        2.0          0.0          0.0   \n",
      "12   YES_complete          1.0        0.0          0.0          1.0   \n",
      "13   YES_complete          2.0        3.0          2.0          3.0   \n",
      "14   YES_complete          0.0        1.0          0.0          2.0   \n",
      "15   YES_complete          3.0        2.0          9.0          3.0   \n",
      "16   YES_complete          2.0        NaN          NaN          2.0   \n",
      "17   YES_complete          2.0        2.0          0.0          1.0   \n",
      "18    YES_only_tu          9.0        NaN          NaN          3.0   \n",
      "19   YES_complete          1.0        0.0          1.0          1.0   \n",
      "20   YES_complete          2.0        3.0          0.0          2.0   \n",
      "21             NO          NaN        2.0          1.0          NaN   \n",
      "22            NaN          NaN        NaN          NaN          NaN   \n",
      "23   YES_complete          1.0        NaN          NaN          1.0   \n",
      "24   YES_complete          1.0        1.0          1.0          2.0   \n",
      "\n",
      "   TILs_tu_LP TILs_raw_ep TILs_raw_tu  \n",
      "0         2.0         2.0         6.0  \n",
      "1         0.0         4.0         7.0  \n",
      "2         3.0        56.0        78.0  \n",
      "3         3.0         5.0        15.0  \n",
      "4         1.0         NaN         NaN  \n",
      "5         NaN         9.0        74.0  \n",
      "6         3.0         2.0         3.0  \n",
      "7         3.0         1.0        11.0  \n",
      "8         2.0         6.0        28.0  \n",
      "9         0.0         NaN         NaN  \n",
      "10        NaN         0.0        22.0  \n",
      "11        2.0         0.0         0.0  \n",
      "12        0.0         2.0         2.0  \n",
      "13        3.0        10.0       116.0  \n",
      "14        1.0         0.0         3.0  \n",
      "15        2.0        32.0       115.0  \n",
      "16        NaN         3.0         4.0  \n",
      "17        2.0        15.0         1.0  \n",
      "18        NaN         NaN        43.0  \n",
      "19        2.0         2.0         2.0  \n",
      "20        3.0         8.0        18.0  \n",
      "21        3.0         NaN         NaN  \n",
      "22        NaN         NaN         NaN  \n",
      "23        NaN         1.0         2.0  \n",
      "24        2.0         1.0         8.0  \n",
      "\n",
      "[25 rows x 56 columns]\n"
     ]
    }
   ],
   "source": [
    "print(final_df.head(25))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
